---
title: "Processing the STAR+HTseq counted RNAseq of batch 2"
author: "Caroline Lohoff"
date: "December 13 2021"
output:
    workflowr::wflow_html:
    toc: true
    theme: cosmo
---

# Set options and load packages

Set global options
```{r setup, include=FALSE, cache = FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
#knitr::opts_knit$set(root.dir = "~/Documents/R/drugseq_test")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load packages
```{r}
library(DESeq2)
library(EnsDb.Hsapiens.v86)
library(RColorBrewer)
library(pheatmap)
library(geneplotter)
library(tidyverse)
```

# Read htseq results
```{r}
rawFolder = "~/Documents/drugSeq/htseq/pilot_batch1old_batch1new_batch2"
fileTab <- tibble(filePath = list.files(rawFolder, full.names = TRUE, recursive = TRUE)) %>%
                    mutate(fileName = basename(filePath)) %>%
                    mutate(fileName = str_remove(fileName,"Aligned_counts.out")) %>% 
  arrange(fileName) %>% mutate(id = paste0("smp",seq(nrow(.))))

countTab <- lapply(seq(nrow(fileTab)), function(i) {
  rec <- fileTab[i,]
  eachTab <- read_delim(rec$filePath, delim = "\t", col_names = FALSE, col_types = "ci") %>%  # 2 columns
    mutate(id = rec$id)
  }) %>% bind_rows() %>%
  pivot_wider(names_from = id, values_from = X2)


countMat <- filter(countTab, !str_detect(X1,"__")) %>%
  data.frame() %>% column_to_rownames("X1") %>%
  as.matrix()

ummapedTab <- filter(countTab, str_detect(X1, "__")) %>%
  pivot_longer(-X1, names_to = "id", values_to = "count") %>%
  mutate(X1 = str_remove_all(X1,"__")) %>%
  dplyr::rename(feature = X1)
```

# Annotate samples
```{r}
load("./data/patmeta_210324.RData")
patAnno <- readxl::read_xlsx("./data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx") %>%
  mutate(fileName = str_replace(fileName, ".txt.gz","")) %>%
  mutate(Replicate = as.factor(Replicate)) %>%
  select(-sequence) %>%
  data.frame(stringsAsFactors = FALSE) %>%
  mutate(diagnosis = patMeta[match(patID, patMeta$Patient.ID),]$diagnosis)

#check all samples have annotations
all(fileTab$fileName %in% patAnno$fileName)
all(patAnno$fileName %in% fileTab$fileName)

#add id column to patAnno
patAnno <- patAnno %>%
  left_join(fileTab, by = "fileName") %>% select(-filePath,-fileName) %>%
  data.frame() %>% column_to_rownames("id")

countMat <- countMat[,colnames(countMat)%in%rownames(patAnno)]  # 412 columns
patAnno <- patAnno[colnames(countMat),]  # 412 rows
```


# Assemble DEseq object

Create ddsDrug object
```{r}
ddsDrug <- DESeqDataSetFromMatrix(countMat, patAnno, design = ~1)
```

Annotate  genes
```{r}
geneAnno <- transcripts(EnsDb.Hsapiens.v86, columns = c("gene_id", "seq_name",
                                                        "gene_name","gene_biotype"), return.type = "DataFrame") %>% as_tibble() %>% distinct(gene_id, .keep_all = TRUE)

geneAnno$tx_id <- NULL

#annotate transcripts
rowAnno <- geneAnno %>% distinct(gene_id, seq_name, gene_name, gene_biotype) %>%
  dplyr::rename(chromosome = seq_name, symbol = gene_name, biotype = gene_biotype) %>%
  data.frame(stringsAsFactors = FALSE) %>% column_to_rownames("gene_id")


rowData(ddsDrug) <- rowAnno[rownames(ddsDrug),]
```

Post processing
```{r}
#remove genes never expressed
ddsDrug <- ddsDrug[rowSums(counts(ddsDrug)) !=0,]

#calculate size factor
ddsDrug <- estimateSizeFactors(ddsDrug)

#add sample ID
ddsDrug$ID <- colnames(ddsDrug)
```

# Summarise htseq statistics
```{r}
mappedTab <- tibble(id = colnames(countMat),
                   feature = "total mapped",
                   count=colSums(countMat))

totalTab <- bind_rows(ummapedTab, mappedTab) %>%
  group_by(id) %>% mutate(total = sum(count)) %>%
  ungroup() %>% mutate(percent = count/total) %>%
  mutate(fileName = fileTab[match(id, fileTab$id),]$fileName)


write_tsv(totalTab, "./output/htseq_process_batch2_20211213.tsv")
```

# Save results
```{r}
save(ddsDrug, file = "./output/ddsDrug_htseq_batch2_20211213.RData")
```

# Session Info Details
```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```
