---
title: "Quality assessment of drug perturbation RNAseq data (batch2 trimmed)"
author: "Caroline Lohoff"
date: "December 13 2021"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Load data and packages

Packages
```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(tidyverse)

knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load data
```{r}
load("~/Documents/R/drugseq_test/output/ddsDrug_trimmed_batch2_20211213.RData")
load("~/Documents/R/drugseq_test/data/patmeta_210324.RData")
```

Global ggplot variables
```{r}
#set the global ggplot theme
theme_set(theme_bw() + theme(axis.text = element_text(size=12), 
                             axis.title = element_text(size=14),
                             plot.title = element_text(size = 15, hjust =0.5, face="bold")))
```

# Distribution of raw counts

## Before normalization
```{r, fig.width=20, fig.height=5}
patAnno <- colData(ddsDrug) %>% as_tibble()
plotTab <- data.frame(counts(ddsDrug)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(batch = patAnno[match(ID, patAnno$ID),]$batch)

ggplot(plotTab, aes(x= ID, y= log10(count), fill = batch)) +
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

```

## After normalization
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsDrug, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(batch = patAnno[match(ID, patAnno$ID),]$batch)

ggplot(plotTab, aes(x= ID, y= log10(count), fill = batch)) +
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```


## After normalization (Batch 1 old and new)
```{r,fig.width=15, fig.height=5}
plotTab_batch1 <- dplyr::filter(plotTab, batch == c("batch1", "batch1old"))

ggplot(plotTab_batch1, aes(x= ID, y= log10(count), fill = batch)) +
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

## After normalization (Pilot and batch 2)
```{r,fig.width=15, fig.height=5}
plotTab_batch2 <- dplyr::filter(plotTab, batch == c("pilot", "batch2"))

ggplot(plotTab_batch2, aes(x= ID, y= log10(count), fill = batch)) +
  geom_boxplot() + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

# Check ECDFs

```{r}
# Remove rows/genes with too few counts
keep <- apply(counts(ddsDrug), 1, function(x) any(x >= 10))
ddsAll <- ddsDrug[keep,]
# Only use protein coding genes
ddsAll <- ddsAll[rowData(ddsAll)$biotype %in% "protein_coding",]
# Remove sex chromosome genes
ddsAll <- ddsAll[!rowData(ddsAll)$chromosome %in% c("X","Y")]
#Variance stabilization transformation of the raw data
RNAnorm.all <- vst(ddsAll)
```

```{r, fig.width=15, fig.height=7, cache=TRUE}
# Successful normalization should lead to overlapped curves
par(mfrow = c(1,2))
geneplotter::multidensity(assay(RNAnorm.all), xlim = c(0,30), legend = F, xlab = "mean counts")
geneplotter::multiecdf(assay(RNAnorm.all), legend = F, xlab="mean counts")
```


# Clustering

## PCA

### Calculate PCA
```{r, fig.height=6, fig.width=8}
#invariant filtering
exprMat <- assay(RNAnorm.all)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:5000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)
```

### PCA colored by batch
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = batch)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
The batches are not separated by PC1 or PC2.


### PCA colored by treatment
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = treatment)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
The treatments are not separated by the principal components. Only the baseline samples of batch1 and batch1old are separated by PC2. 

### PCA colored by patientID
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = patID)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
### PCA colored by sampleID
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = sampleID)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```

## Sample distance
```{r, fig.width=20, fig.height=20}
sampleDists <- dist(t(assay(RNAnorm.all)))
sampleDistMat <- as.matrix(sampleDists)
annoCol <- patAnno %>% select(expID, expBatch, patID, FSC.SSC.norm, treatment, time, batch, ID) %>%
  data.frame() %>% column_to_rownames("ID")
#Plot heatmap
colors <- colorRampPalette( rev(brewer.pal(9,"Blues")) )(255)
pheatmap(sampleDistMat, color = colors, clustering_distance_rows = sampleDists, clustering_distance_cols = sampleDists, annotation_col = annoCol, clustering_method = "ward.D2",
         labels_col = paste0(patAnno[colnames(sampleDistMat),]$patID,"_",
                             patAnno[colnames(sampleDistMat),]$treatment))
```


## Sample similarity
```{r, fig.width=20, fig.height=20}
sampleSimMat <- cor(assay(RNAnorm.all))
pheatmap(sampleSimMat, annotation_col = annoCol,
          clustering_method = "ward.D2",
         labels_col = paste0(patAnno[colnames(sampleSimMat),]$patID,"_",
                             patAnno[colnames(sampleSimMat),]$treatment))
```


# Adjusting for patient specific effect
The technical outliers should be dealt with before adjusting for patient-specific effect.


# Compare with previous RNAseq data (samples with baseline included)

## Processing datasets
```{r}
load("~/Documents/R/drugseq_test/data/ddsCLL_Salmon_190702.RData")
ddsSub <- ddsDrug[,ddsDrug$treatment %in% "Baseline"]   # 9 samples in new RNA seq data contain baseline
overPat <- intersect(ddsCLL$patientID, ddsSub$patID)    # 5 overlapping patient IDs
ddsOld <- ddsCLL[,overPat]  
ddsNew <- ddsDrug[,match(overPat, ddsDrug$patID)]  

#only keep protein coding genes
ddsOld <- ddsOld[rowData(ddsOld)$gene_biotype %in% "protein_coding",]  # dim 22028 rows x 5 columns
ddsNew <- ddsNew[rowData(ddsNew)$biotype %in% "protein_coding",]       # dim 22028 rows x 5 columns
colnames(ddsNew) <- ddsNew$patID
```
5 overlapping patient IDs between old (2019) and new RNAseq (2021) data.

## Summarise detected genes
```{r, warning=FALSE, message=FALSE}
sumCountTab <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  geneOld <- rownames(subOld)[rowSums(assay(subOld)) > 0]
  geneNew <- rownames(subNew)[rowSums(assay(subNew)) > 0]
  tibble(patID = n,
         onlyNew = length(setdiff(geneNew, geneOld)),
         onlyOld = length(setdiff(geneOld, geneNew)),
         both = length(intersect(geneNew, geneOld)))
}) %>% bind_rows()
sumCountTab
```

```{r}
plotTab <- sumCountTab %>% pivot_longer(-patID, names_to = "group", values_to = "count") %>%
  mutate(group = factor(group, level = c("onlyNew","both","onlyOld")))
ggplot(plotTab, aes(x=patID, y = count, fill = group)) + geom_bar(stat = "identity") +
    theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
```

## Correlations of gene counts

```{r, fig.width=10, fig.height=10, message=FALSE, warning=FALSE}
plotList <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  overGene <- intersect(rownames(subOld), rownames(subNew))
  subOld <- subOld[overGene,]
  subNew <- subNew[overGene,]
  plotTab <- tibble(geneID = overGene,
                    countOld = assay(subOld)[,1],
                    countNew = assay(subNew)[,1])
  ggplot(plotTab, aes(x=log10(countOld), y=log10(countNew))) + scattermore::geom_scattermore() +
     ggtitle(n)
})
cowplot::plot_grid(plotList[[1]], plotList[[2]], plotList[[3]], plotList[[4]], plotList[[5]], ncol=3, labels="AUTO")
```


## Compare the similarity of expression pattern

Assemble new combined DESeq object
```{r}
oldMat <- counts(ddsOld)
newMat <- counts(ddsNew)
colnames(newMat) <- paste0(colnames(newMat),"_1")
patBack <- data.frame(row.names = c(colnames(oldMat), colnames(newMat)),
                      patID = c(ddsOld$patientID,ddsNew$patID),
                      set = rep(c("old","new"), each = ncol(oldMat)))
overGene <- intersect(rownames(oldMat), rownames(newMat))
comMat <- cbind(oldMat[overGene,],newMat[overGene,])
ddsCom <- DESeqDataSetFromMatrix(comMat, patBack, design= ~1)
ddsCom <- estimateSizeFactors(ddsCom)  #estimate size factor
```

Distribution after normalization
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsCom, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(set = patBack[ID,]$set)

ggplot(plotTab, aes(x= ID, y= log10(count))) +
  geom_boxplot(aes(fill = set)) + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

Variance stabilization transformation
```{r}
ddsCom.vst <- varianceStabilizingTransformation(ddsCom)
```

### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.vst)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

# create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```
The batch (or sequencing platform) explains the major difference

### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol, clustering_method = "ward.D2")
```


## Compare similarity of expression pattern after adjusting for batch

Adjust batch effect using limma
```{r}
exprMat <- assay(ddsCom.vst)
exprMat <- sva::ComBat(exprMat, batch = factor(ddsCom.vst$set))
ddsCom.adj <- ddsCom.vst
assay(ddsCom.adj) <- exprMat
```


### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom.adj) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.adj)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```


### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol, clustering_method = "ward.D2")
```


# Compare with previous RNAseq data (using DMSO as baseline)

## Processing datasets
```{r}
load("~/Documents/R/drugseq_test/data/ddsCLL_Salmon_190702.RData")
ddsSub <- ddsDrug[,ddsDrug$treatment %in% "DMSO"]    # 38 samples were treated with DMSO
overPat <- intersect(ddsCLL$patientID, ddsSub$patID) # 31 overlapping patient IDs
ddsOld <- ddsCLL[,overPat]  
ddsNew <- ddsDrug[,match(overPat, ddsDrug$patID)] 

#only keep protein coding genes
ddsOld <- ddsOld[rowData(ddsOld)$gene_biotype %in% "protein_coding",] # dim: 22028 rows x 31 columns
ddsNew <- ddsNew[rowData(ddsNew)$biotype %in% "protein_coding",]      # dim: 22028 rows x 31 columns
colnames(ddsNew) <- ddsNew$patID
```
31 patient IDs treated with DMSO are shared between old (2019) and new RNAseq (2021) data.


## Summarise detected genes
```{r}
sumCountTab <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  geneOld <- rownames(subOld)[rowSums(assay(subOld)) > 0]
  geneNew <- rownames(subNew)[rowSums(assay(subNew)) > 0]
  tibble(patID = n,
         onlyNew = length(setdiff(geneNew, geneOld)),
         onlyOld = length(setdiff(geneOld, geneNew)),
         both = length(intersect(geneNew, geneOld)))
}) %>% bind_rows()
sumCountTab
```

```{r}
plotTab <- sumCountTab %>% pivot_longer(-patID, names_to = "group", values_to = "count") %>%
  mutate(group = factor(group, level = c("onlyNew","both","onlyOld")))
ggplot(plotTab, aes(x=patID, y = count, fill = group)) + geom_bar(stat = "identity")
```
The overlap of the detected genes look good.
**both**: genes have non-zero count in both datasets
**onlyNew**: genes detected only in the current batch 1 new from 2021 (shallow sequencing)
**onlyOld**: genes detected only in the previous batch from 2019 (full depth)  


## Compare the similarity of expression pattern

Assemble new combined DESeq object
```{r}
oldMat <- counts(ddsOld)
newMat <- counts(ddsNew)
colnames(newMat) <- paste0(colnames(newMat),"_1")
patBack <- data.frame(row.names = c(colnames(oldMat), colnames(newMat)),
                      patID = c(ddsOld$patientID,ddsNew$patID),
                      set = rep(c("old","new"), each = ncol(oldMat)))
overGene <- intersect(rownames(oldMat), rownames(newMat))
comMat <- cbind(oldMat[overGene,],newMat[overGene,])
ddsCom <- DESeqDataSetFromMatrix(comMat, patBack, design= ~1)
ddsCom <- estimateSizeFactors(ddsCom)  #estimate size factor
```

Distribution after normalization
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsCom, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(set = patBack[ID,]$set)

ggplot(plotTab, aes(x= ID, y= log10(count))) +
  geom_boxplot(aes(fill = set)) + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

Variance stabilization transformation
```{r}
ddsCom.vst <- varianceStabilizingTransformation(ddsCom)
```

### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.vst)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```
The batch (or sequencing platform) explains the major difference.

### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol,
          clustering_method = "ward.D2")
```


## Compare similarity of expression pattern after adjusting for batch

Adjust batch effect using limma
```{r}
exprMat <- assay(ddsCom.vst)
exprMat <- limma::removeBatchEffect(exprMat, batch = factor(ddsCom.vst$set))
ddsCom.adj <- ddsCom.vst
assay(ddsCom.adj) <- exprMat
```


### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom.adj) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.adj)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```


### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol,
          clustering_method = "ward.D2")
```
