<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Caroline Lohoff" />


<title>Check alignment quality by STAR of batch 2</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="site_libs/datatables-binding-0.20/datatables.js"></script>
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="site_libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="site_libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">drugseq_test</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Check alignment quality by STAR of batch 2</h1>
<h4 class="author">Caroline Lohoff</h4>
<h4 class="date">December 13 2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-12-13
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>drugseq_test/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20211129code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20211129)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20211129code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20211129)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongabsolute"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>File paths:</strong> absolute </a>
</p>
</div>
<div id="strongFilepathsstrongabsolute" class="panel-collapse collapse">
<div class="panel-body">
<p>
Using absolute paths to the files within your workflowr project makes it difficult for you and others to run your code on a different machine. Change the absolute path(s) below to the suggested relative path(s) to make your code more reproducible.
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
absolute
</th>
<th style="text-align:left;">
relative
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
~/Documents/R/drugseq_test/data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx
</td>
<td style="text-align:left;">
../data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx
</td>
</tr>
<tr>
<td style="text-align:left;">
~/Documents/R/drugseq_test/output/htseq_process_batch2_20211213.tsv
</td>
<td style="text-align:left;">
../output/htseq_process_batch2_20211213.tsv
</td>
</tr>
<tr>
<td style="text-align:left;">
~/Documents/R/drugseq_test/output/STAR_HTseq_batch2_summary_20211213.csv
</td>
<td style="text-align:left;">
../output/STAR_HTseq_batch2_summary_20211213.csv
</td>
</tr>
<tr>
<td style="text-align:left;">
~/Documents/R/drugseq_test/data/patmeta_210324.RData
</td>
<td style="text-align:left;">
../data/patmeta_210324.RData
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgitembldejludrugseqtree4039c41558a49e9634df60a08e1bca8b2b112cdbtargetblank4039c41a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://git.embl.de/jlu/drugseq/tree/4039c41558a49e9634df60a08e1bca8b2b112cdb" target="_blank">4039c41</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgitembldejludrugseqtree4039c41558a49e9634df60a08e1bca8b2b112cdbtargetblank4039c41a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://git.embl.de/jlu/drugseq/tree/4039c41558a49e9634df60a08e1bca8b2b112cdb" target="_blank">4039c41</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store
    Ignored:    data/.DS_Store

Untracked files:
    Untracked:  analysis/1_process_batch1new_trimmed.Rmd
    Untracked:  analysis/1_process_batch1new_untrimmed.Rmd
    Untracked:  analysis/1_process_batch2_trimmed.Rmd
    Untracked:  analysis/1_process_batch2_untrimmed.Rmd
    Untracked:  analysis/1_process_htseq_batch1new.Rmd
    Untracked:  analysis/1_process_htseq_batch2.Rmd
    Untracked:  analysis/2_qualityCheck_Batch2_trimmed.Rmd
    Untracked:  analysis/2_qualityCheck_Batch2_untrimmed.Rmd
    Untracked:  analysis/2_qualityCheck_HTseq_batch1new.Rmd
    Untracked:  analysis/2_qualityCheck_HTseq_batch2.Rmd
    Untracked:  analysis/2_qualityCheck_bam_batch1new.Rmd
    Untracked:  analysis/2_qualityCheck_bam_batch2.Rmd
    Untracked:  analysis/2_qualityCheck_pilotAndBatch1new_trimmed.Rmd
    Untracked:  analysis/2_qualityCheck_pilotAndBatch1new_untrimmed.Rmd
    Untracked:  analysis/3_differentialExpression_batch1new_HTseq.Rmd
    Untracked:  analysis/3_differentialExpression_batch1new_trimmed.Rmd
    Untracked:  analysis/3_differentialExpression_batch1new_untrimmed.Rmd
    Untracked:  analysis/3_differentialExpression_batch2_untrimmed.Rmd
    Untracked:  analysis/4_fastQC_batch1new.Rmd
    Untracked:  analysis/Camera_function_limma.R
    Untracked:  analysis/figure/
    Untracked:  code/Changes_in_RNAseq_samples_all.R
    Untracked:  code/genAlign_salmon_QuantSeq_batch1new_trimmed.R
    Untracked:  code/genAlign_salmon_QuantSeq_batch1new_untrimmed.R
    Untracked:  code/genAlign_salmon_QuantSeq_batch2_trimmed.R
    Untracked:  code/genAlign_salmon_QuantSeq_batch2_untrimmed.R
    Untracked:  code/genAlign_star_QuantSeq_batch1new.R
    Untracked:  code/genAlign_star_QuantSeq_batch2.R
    Untracked:  code/genCount_batch1new.R
    Untracked:  code/genCount_batch2.R
    Untracked:  code/genFastQC_batch1new.R
    Untracked:  code/genFastQC_batch2.R
    Untracked:  data/4_differentialExpression.RData
    Untracked:  data/RNAseq_samples_pilot_batch1new.xlsx
    Untracked:  data/RNAseq_samples_pilot_batch1new_batch2.xlsx
    Untracked:  data/RNAseq_samples_pilot_batch1old.xlsx
    Untracked:  data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx
    Untracked:  data/bamList_batch1.txt
    Untracked:  data/bamList_batch1new.txt
    Untracked:  data/bamList_batch1new_DMSO.txt
    Untracked:  data/bamList_batch2.txt
    Untracked:  data/bamList_pilot.txt
    Untracked:  data/bamList_pilot_DMSO.txt
    Untracked:  data/c2.cp.kegg.v6.2.symbols.gmt.txt
    Untracked:  data/ddsCLL_Salmon_190702.RData
    Untracked:  data/ddsrna_180717.RData
    Untracked:  data/drosphila_example_de.csv
    Untracked:  data/fastqFiles_batch1.txt
    Untracked:  data/fastqFiles_batch1new.txt
    Untracked:  data/fastqFiles_batch2.txt
    Untracked:  data/fastqFiles_pilot.txt
    Untracked:  data/h.all.v6.2.symbols.gmt.txt
    Untracked:  data/human_c6_v5p2.rdata
    Untracked:  data/patmeta_210324.RData
    Untracked:  data/~$RNAseq_samples_all.xlsx
    Untracked:  data/~$RNAseq_samples_all_batch1new.xlsx
    Untracked:  differentialExpression.RData
    Untracked:  drugseq_test.Rproj
    Untracked:  output/

Unstaged changes:
    Modified:   _workflowr.yml
    Modified:   analysis/_site.yml
    Deleted:    analysis/bamQC.Rmd
    Deleted:    analysis/bamQC_HAP1.Rmd
    Deleted:    analysis/calcCoverage.Rmd
    Deleted:    analysis/differentialExpression_batch1.Rmd
    Deleted:    analysis/differentialExpression_batch1_HTseq.Rmd
    Deleted:    analysis/process.Rmd
    Deleted:    analysis/process_htseq.Rmd
    Deleted:    analysis/qualityCheck_HTseq.Rmd
    Deleted:    analysis/qualityCheck_pilotAndBatch1.Rmd
    Deleted:    analysis/qualityCheck_pilotAndBatch1_trimmed.Rmd
    Deleted:    code/fastqFiles.txt
    Deleted:    code/genAlign_salmon_QuantSeq.R
    Deleted:    code/genAlign_star_QuantSeq.R
    Deleted:    code/genCount.R
    Deleted:    code/genFastQC.R
    Deleted:    data/RNAseq_samples_all.xlsx
    Deleted:    drugseq.Rproj

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
<pre><code>✓ ggplot2 3.3.5     ✓ purrr   0.3.4
✓ tibble  3.1.6     ✓ dplyr   1.0.7
✓ tidyr   1.1.4     ✓ stringr 1.4.0
✓ readr   2.1.0     ✓ forcats 0.5.1</code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()</code></pre>
<div id="quality-information-from-star-alignment-genome-mapping" class="section level1">
<h1>Quality information from STAR alignment (genome mapping)</h1>
<pre class="r"><code>processLog &lt;- function(fileName) {
    outTab &lt;- tibble(allLine = read_lines(fileName)) %&gt;%
      separate(allLine, c(&quot;entry&quot;,&quot;value&quot;),&quot;[|]&quot;, remove = FALSE) %&gt;%
      mutate(value = str_remove_all(value, &quot;[\t%]&quot;),
             entry = str_trim(entry,&quot;both&quot;)) %&gt;%
      mutate(value = as.numeric(value)) %&gt;%
      filter(!is.na(value)) %&gt;% select(-allLine)
    
    outTab
}</code></pre>
</div>
<div id="load-bam-files-from-pilot-and-batch-1-new" class="section level1">
<h1>Load bam files from pilot and batch 1 new</h1>
<pre class="r"><code>allFiles &lt;- list.files(&quot;~/Documents/drugSeq/bam/pilot_batch1old_batch1new_batch2/log/&quot;,recursive = TRUE, full.names = TRUE)
allRecord &lt;- lapply(allFiles, function(fileName){
  baseName &lt;- basename(fileName)
  processLog(fileName) %&gt;%
    mutate(filename = str_remove(baseName, &quot;Log.final.out&quot;))
}) %&gt;% bind_rows()</code></pre>
<p>Annotate patients</p>
<pre class="r"><code>patAnno &lt;- readxl::read_xlsx(&quot;~/Documents/R/drugseq_test/data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx&quot;) %&gt;%
  mutate(fileName = str_replace(fileName, &quot;.txt.gz&quot;,&quot;&quot;)) %&gt;%
  mutate(Replicate = as.factor(Replicate)) %&gt;%
  select(-sequence) %&gt;%
  mutate(cell = ifelse(patID == &quot;Jurkat&quot;,&quot;Jurkat&quot;,&quot;Primary&quot;)) %&gt;%
  select(fileName, expBatch, patID, sampleID, batch, cell)

#check all samples have annotations
all(allRecord$filename %in% patAnno$fileName)</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>all(patAnno$fileName %in% allRecord$filename) </code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>allRecord &lt;- left_join(allRecord, patAnno, by = c(filename = &quot;fileName&quot;)) %&gt;% arrange(entry)</code></pre>
<pre class="r"><code>diagBoxPlots &lt;- lapply(unique(allRecord$entry), function(n) {
  plotTab &lt;- filter(allRecord, entry == n)
  ggplot(plotTab, aes(x=batch, y=value)) +
    geom_boxplot(outlier.shape = NA) +
    ggbeeswarm::geom_quasirandom(aes(col=cell), groupOnX=TRUE) +
    theme_bw() + theme(legend.position = &quot;none&quot;) +
    ggtitle(n)
})
cowplot::plot_grid(plotlist = diagBoxPlots, ncol=3)</code></pre>
<p><img src="figure/2_qualityCheck_bam_batch2.Rmd/unnamed-chunk-5-1.png" width="1440" style="display: block; margin: auto;" /></p>
</div>
<div id="quality-information-from-htseq-gene-counting" class="section level1">
<h1>Quality information from HTseq (gene counting)</h1>
<pre class="r"><code>htTab &lt;- read_tsv(&quot;~/Documents/R/drugseq_test/output/htseq_process_batch2_20211213.tsv&quot;)</code></pre>
<pre><code>Rows: 6108 Columns: 6</code></pre>
<pre><code>── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr (3): feature, id, fileName
dbl (3): count, total, percent</code></pre>
<pre><code>
ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>htTab &lt;- htTab %&gt;% mutate(batch = patAnno[match(fileName, patAnno$fileName),]$batch,
                          cell = patAnno[match(fileName, patAnno$fileName),]$cell) %&gt;%
  mutate(percent = percent*100)</code></pre>
<pre class="r"><code>diagBoxPlots &lt;- lapply(unique(htTab$feature), function(n) {
  plotTab &lt;- filter(htTab, feature == n)
  ggplot(plotTab, aes(x=batch, y=percent)) +
    geom_boxplot(outlier.shape = NA) +
    ggbeeswarm::geom_quasirandom(aes(col = cell), groupOnX=TRUE) +
    ylab(&quot;Percentage&quot;) +
    theme_bw() + theme(legend.position = &quot;none&quot;) +
    ggtitle(n)
})
cowplot::plot_grid(plotlist = diagBoxPlots, ncol=3)</code></pre>
<p><img src="figure/2_qualityCheck_bam_batch2.Rmd/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Combine and save</p>
<pre class="r"><code>htTabReformated &lt;- select(htTab, feature, count, percent, fileName) %&gt;%
  pivot_longer(c(count, percent), names_to = &quot;type&quot;, values_to = &quot;value&quot;) %&gt;%
  mutate(entry = ifelse(type == &quot;count&quot;, feature, paste0(feature,&quot; %&quot;))) %&gt;%
  select(value, entry, fileName) %&gt;%
  left_join(select(patAnno, fileName, expBatch, patID, sampleID, batch), by = &quot;fileName&quot;) %&gt;%
  mutate(source = &quot;HTseq&quot;) %&gt;%
  dplyr::rename(filename  = fileName)

allRecordOut &lt;- bind_rows(allRecord %&gt;% mutate(source = &quot;STAR&quot;),
                          htTabReformated)

write_csv2(allRecordOut, file = &quot;~/Documents/R/drugseq_test/output/STAR_HTseq_batch2_summary_20211213.csv&quot;)</code></pre>
</div>
<div id="salmon-results" class="section level1">
<h1>Salmon results</h1>
<p>Pilot batch</p>
<pre class="r"><code>allDir &lt;- list.dirs(&quot;~/Documents/drugSeq/salmon/pilot/&quot;, recursive = FALSE)
salmonOut.pilot &lt;- lapply(allDir, function(n) {
  filename &lt;- basename(n)
  allLines &lt;- read_lines(paste0(n,&quot;/logs/salmon_quant.log&quot;))
  allLines &lt;- allLines[str_detect(allLines, &quot;Mapping rate =&quot;)]
  mapRate &lt;- str_split(allLines, &quot;=&quot;)[[1]][2] %&gt;% str_remove_all(&quot;[ %]&quot;) %&gt;% as.numeric()
  tibble(filename = filename, value = mapRate)
}) %&gt;% bind_rows() %&gt;%
  left_join(patAnno, by = c(filename = &quot;fileName&quot;))</code></pre>
<p>Batch 1 old</p>
<pre class="r"><code>allDir &lt;- list.dirs(&quot;~/Documents/drugSeq/salmon/batch1/&quot;, recursive = FALSE)
salmonOut.batch1old &lt;- lapply(allDir, function(n) {
  filename &lt;- basename(n)
  allLines &lt;- read_lines(paste0(n,&quot;/logs/salmon_quant.log&quot;))
  allLines &lt;- allLines[str_detect(allLines, &quot;Mapping rate =&quot;)]
  mapRate &lt;- str_split(allLines, &quot;=&quot;)[[1]][2] %&gt;% str_remove_all(&quot;[ %]&quot;) %&gt;% as.numeric()
  tibble(filename = filename, value = mapRate)
}) %&gt;% bind_rows() %&gt;%
  left_join(patAnno, by = c(filename = &quot;fileName&quot;))</code></pre>
<p>Batch 1 new</p>
<pre class="r"><code>allDir &lt;- list.dirs(&quot;~/Documents/drugSeq/salmon/batch1_new/&quot;, recursive = FALSE)
salmonOut.batch1new &lt;- lapply(allDir, function(n) {
  filename &lt;- basename(n)
  allLines &lt;- read_lines(paste0(n,&quot;/logs/salmon_quant.log&quot;))
  allLines &lt;- allLines[str_detect(allLines, &quot;Mapping rate =&quot;)]
  mapRate &lt;- str_split(allLines, &quot;=&quot;)[[1]][2] %&gt;% str_remove_all(&quot;[ %]&quot;) %&gt;% as.numeric()
  tibble(filename = filename, value = mapRate)
}) %&gt;% bind_rows() %&gt;%
  left_join(patAnno, by = c(filename = &quot;fileName&quot;))</code></pre>
<p>Batch 2</p>
<pre class="r"><code>allDir &lt;- list.dirs(&quot;~/Documents/drugSeq/salmon/batch2/&quot;, recursive = FALSE)
salmonOut.batch2 &lt;- lapply(allDir, function(n) {
  filename &lt;- basename(n)
  allLines &lt;- read_lines(paste0(n,&quot;/logs/salmon_quant.log&quot;))
  allLines &lt;- allLines[str_detect(allLines, &quot;Mapping rate =&quot;)]
  mapRate &lt;- str_split(allLines, &quot;=&quot;)[[1]][2] %&gt;% str_remove_all(&quot;[ %]&quot;) %&gt;% as.numeric()
  tibble(filename = filename, value = mapRate)
}) %&gt;% bind_rows() %&gt;%
  left_join(patAnno, by = c(filename = &quot;fileName&quot;))</code></pre>
</div>
<div id="plot-mapping-rate" class="section level1">
<h1>Plot mapping rate</h1>
<pre class="r"><code>salmonOut &lt;- bind_rows(salmonOut.pilot, salmonOut.batch1old, salmonOut.batch1new, salmonOut.batch2)
ggplot(salmonOut, aes(x=batch, y=value)) +
  geom_boxplot(outlier.shape = NA) +
  ggbeeswarm::geom_quasirandom(aes(col = cell)) +
  ylab(&quot;Percentage&quot;)  +
  theme_bw() + theme(legend.position = &quot;none&quot;) +
  ggtitle(&quot;Salmon map rate&quot;)</code></pre>
<p><img src="figure/2_qualityCheck_bam_batch2.Rmd/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;" /></p>
</div>
<div id="correlation-quality-features-with-cell-viability" class="section level1">
<h1>Correlation quality features with cell viability</h1>
<pre class="r"><code>load(&quot;~/Documents/R/drugseq_test/data/patmeta_210324.RData&quot;)

allRecordOut &lt;- read_csv2(&quot;~/Documents/R/drugseq_test/output/STAR_HTseq_batch2_summary_20211213.csv&quot;) %&gt;%
  filter(str_detect(entry,&quot;%&quot;), batch == &quot;batch1&quot;) %&gt;%
  select(entry, value, filename)</code></pre>
<pre><code>ℹ Using &quot;&#39;,&#39;&quot; as decimal and &quot;&#39;.&#39;&quot; as grouping mark. Use `read_delim()` for more control.</code></pre>
<pre><code>Rows: 41854 Columns: 9</code></pre>
<pre><code>── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;;&quot;
chr (8): entry, filename, expBatch, patID, sampleID, batch, cell, source
dbl (1): value</code></pre>
<pre><code>
ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>patAnno &lt;- readxl::read_xlsx(&quot;~/Documents/R/drugseq_test/data/RNAseq_samples_pilot_batch1old_batch1new_batch2.xlsx&quot;) %&gt;%
  mutate(fileName = str_replace(fileName, &quot;.txt.gz&quot;,&quot;&quot;)) %&gt;%
  mutate(diagnosis = patMeta[match(patID, patMeta$Patient.ID),]$diagnosis) %&gt;%
  mutate(diagnosis = ifelse(patID == &quot;Jurkat&quot;,&quot;Jurkat&quot;,diagnosis))

patTest &lt;- patAnno %&gt;%
  select(fileName, FSC.SSC.d0, FSC.SSC, FSC.SSC.norm) %&gt;%
  pivot_longer(-fileName, names_to = &quot;metaVar&quot;, values_to = &quot;metaValue&quot;)</code></pre>
<pre class="r"><code>testTab &lt;- full_join(allRecordOut, patTest, by = c(filename= &quot;fileName&quot;)) %&gt;%
  filter(!is.na(value), !is.na(metaValue))
  
resTab &lt;- group_by(testTab, entry, metaVar) %&gt;% nest() %&gt;%
  mutate(m=map(data, ~cor.test(~value + metaValue, .,
                              method = &quot;spearman&quot;, use = &quot;pairwise.complete.obs&quot;))) %&gt;%
  mutate(res = map(m, broom::tidy)) %&gt;% 
  unnest(res) %&gt;%
  select(entry, metaVar, p.value, estimate) %&gt;%
  arrange(p.value)

resTab %&gt;% mutate_if(is.numeric, formatC) %&gt;%
  DT::datatable()</code></pre>
<div id="htmlwidget-d6d993e391e6facc9610" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-d6d993e391e6facc9610">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42"],["ambiguous %","total mapped %","% of reads unmapped: other","% of reads mapped to too many loci","% of reads unmapped: too short","% of reads mapped to multiple loci","alignment_not_unique %","Uniquely mapped reads %","total mapped %","Mismatch rate per base, %","no_feature %","ambiguous %","no_feature %","% of reads mapped to too many loci","% of reads unmapped: too short","% of reads unmapped: other","% of reads unmapped: other","no_feature %","Mismatch rate per base, %","% of reads mapped to multiple loci","% of reads mapped to too many loci","Mismatch rate per base, %","% of reads mapped to multiple loci","% of reads unmapped: too short","alignment_not_unique %","Uniquely mapped reads %","alignment_not_unique %","total mapped %","Uniquely mapped reads %","ambiguous %","% of chimeric reads","% of chimeric reads","% of chimeric reads","% of reads unmapped: too many mismatches","% of reads unmapped: too many mismatches","% of reads unmapped: too many mismatches","too_low_aQual %","too_low_aQual %","too_low_aQual %","not_aligned %","not_aligned %","not_aligned %"],["FSC.SSC","FSC.SSC","FSC.SSC","FSC.SSC","FSC.SSC","FSC.SSC.d0","FSC.SSC.d0","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC","FSC.SSC","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC.norm","FSC.SSC","FSC.SSC.d0","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC.norm","FSC.SSC","FSC.SSC","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC.d0","FSC.SSC","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC","FSC.SSC.norm","FSC.SSC.d0","FSC.SSC","FSC.SSC.norm"],["2.471e-28","2.148e-27","7.663e-21","9.745e-17","1.774e-14","2.575e-10","6.671e-10","8.809e-10","1.071e-09","1.289e-09","1.793e-09","2.609e-09","2.129e-08","2.953e-07","3.375e-07","6.884e-07","0.0001118","0.001981","0.003197","0.03626","0.2738","0.2864","0.3721","0.3853","0.4169","0.6067","0.7213","0.7408","0.7514","0.9691","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA"],["0.5356","0.5275","-0.463","-0.4159","-0.3864","0.3261","0.3201","-0.3167","0.3131","-0.3104","-0.309","0.3061","-0.2918","-0.2641","-0.2657","-0.256","-0.2028","-0.1618","-0.1537","0.1095","-0.05799","0.0565","0.04679","0.04551","0.04274","-0.027","0.01879","-0.01762","-0.01661","0.002064","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA","NA"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>entry<\/th>\n      <th>metaVar<\/th>\n      <th>p.value<\/th>\n      <th>estimate<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script>
</div>
<div id="top-9-associations-as-scatter-plot" class="section level1">
<h1>Top 9 associations as scatter plot</h1>
<pre class="r"><code>pList &lt;- lapply(seq(12), function(i) {
  rec &lt;- resTab[i,]
  plotTab &lt;- filter(testTab, entry == rec$entry, metaVar == rec$metaVar) %&gt;%
    mutate(diagnosis = patAnno[match(filename, patAnno$fileName),]$diagnosis)
  ggplot(plotTab, aes(x=metaValue, y=value)) +
    geom_point(aes(col = diagnosis)) + geom_smooth(method = &quot;lm&quot;) +
    theme_bw() +
    xlab(rec$metaVar) + ylab(rec$entry)
})
cowplot::plot_grid(plotlist = pList, ncol=3)</code></pre>
<p><img src="figure/2_qualityCheck_bam_batch2.Rmd/unnamed-chunk-16-1.png" width="1152" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.2 (2021-11-01)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7     purrr_0.3.4    
[5] readr_2.1.0     tidyr_1.1.4     tibble_3.1.6    ggplot2_3.3.5  
[9] tidyverse_1.3.1

loaded via a namespace (and not attached):
 [1] httr_1.4.2        sass_0.4.0        splines_4.1.2     bit64_4.0.5      
 [5] vroom_1.5.6       jsonlite_1.7.2    modelr_0.1.8      bslib_0.3.1      
 [9] assertthat_0.2.1  highr_0.9         vipor_0.4.5       cellranger_1.1.0 
[13] yaml_2.2.1        lattice_0.20-45   pillar_1.6.4      backports_1.4.0  
[17] glue_1.5.1        digest_0.6.29     promises_1.2.0.1  rvest_1.0.2      
[21] colorspace_2.0-2  Matrix_1.3-4      cowplot_1.1.1     htmltools_0.5.2  
[25] httpuv_1.6.3      pkgconfig_2.0.3   broom_0.7.10      haven_2.4.3      
[29] scales_1.1.1      later_1.3.0       tzdb_0.2.0        git2r_0.29.0     
[33] mgcv_1.8-38       generics_0.1.1    farver_2.1.0      ellipsis_0.3.2   
[37] DT_0.20           withr_2.4.3       cli_3.1.0         magrittr_2.0.1   
[41] crayon_1.4.2      readxl_1.3.1      evaluate_0.14     fs_1.5.0         
[45] fansi_0.5.0       nlme_3.1-153      xml2_1.3.2        beeswarm_0.4.0   
[49] tools_4.1.2       hms_1.1.1         lifecycle_1.0.1   munsell_0.5.0    
[53] reprex_2.0.1      compiler_4.1.2    jquerylib_0.1.4   rlang_0.4.12     
[57] grid_4.1.2        rstudioapi_0.13   htmlwidgets_1.5.4 crosstalk_1.2.0  
[61] labeling_0.4.2    rmarkdown_2.11    gtable_0.3.0      DBI_1.1.1        
[65] R6_2.5.1          lubridate_1.8.0   knitr_1.36        fastmap_1.1.0    
[69] bit_4.0.4         utf8_1.2.2        workflowr_1.6.2   rprojroot_2.0.2  
[73] stringi_1.7.5     ggbeeswarm_0.6.0  parallel_4.1.2    Rcpp_1.0.7       
[77] vctrs_0.3.8       dbplyr_2.1.1      tidyselect_1.1.1  xfun_0.28        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
