---
title: "Processing the STAR+HTseq counted RNAseq of batch 4"
author: "Caroline Lohoff (base on code from Junyan Lu)"
date: "February 07 2022"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Description
The aim is to process RNA-Sequencing data from patients with chronic lymphocytic leukemia (CLL). The untrimmed RNA-Seq reads after drug perturbation were pre-processed with [STAR](https://doi.org/10.1093/bioinformatics/bts635) version 2.7.9a by aligning the sequence reads to the Ensembl human reference genome (GRCh38) and creating a list of genomic features. STAR was run in default mode without adapter trimming. The output was saved as BAM files and subsequently used to count how many reads map to each gene with the tool htseq-count version 0.9.1 from the [HTSeq](https://doi.org/10.1093/bioinformatics/btu638) python package with union mode. Thus, only reads unambiguously mapping to a single gene were counted. For RNA-Seq each gene is typically considered as the union of all its exons. The output of HTSeq is a table with counts for each feature that is well-suited for differential expression analysis.
This RMD file imports and gathers the gene counts from all batches (pilot, batch 1 combined, batch 2, batch 3, and batch 4) and creates a count matrix. The read counts together with patient annotations are saved as a DESeq2 object. 

# Set options and load packages

Set global options
```{r setup, include=FALSE, cache = FALSE}
#knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_knit$set(root.dir = "~/Documents/R/drugseq_test")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load packages
```{r load-packages}
suppressPackageStartupMessages({
  library(DESeq2)
  library(EnsDb.Hsapiens.v86)
  library(tidyverse)
})
```

# Load patient annotations
```{r annotation}
# Read in annotation file
patAnno <- readxl::read_xlsx("./data/RNAseq_samples_pilot_batch1-4_20220207.xlsx") %>%
  mutate(fileName = str_replace(fileName, ".txt.gz","")) %>%
  mutate(Replicate = as.factor(Replicate)) %>%
  dplyr::select(-sequence) %>%
  data.frame(stringsAsFactors = FALSE) %>%
  mutate(ID = paste0("smp",seq(nrow(.))))

# Create subset of data frame for joining with fileTab
patAnno_small <- patAnno[c("fileName", "ID")]
patAnno_small <- arrange(patAnno_small, fileName) %>%
  mutate(IDplate = paste0("fn",seq(nrow(.))))

# Add column to patAnno
patAnno_join <- patAnno_small[c("fileName", "IDplate")]
patAnno <- patAnno %>% left_join(patAnno_join, by = "fileName")
```

# Read htseq results and create a count matrix
```{r create-matrix}
# Read in count tables of pilot batch, batch 1 combined, and batches 2-4
rawFolder = "~/Documents/drugSeq/htseq/pilot_batch1comb_batch2_batch3_batch4"
fileTab <- tibble(filePath = list.files(rawFolder, full.names = TRUE,
                                        recursive = TRUE)) %>%
                    mutate(fileName = basename(filePath)) %>%
                    mutate(fileName = str_remove(fileName,
                                                 "Aligned_counts.out")) %>%
                    left_join(patAnno_small, by = "fileName") %>%
                    dplyr::select(-IDplate)

countTab <- lapply(seq(nrow(fileTab)), function(i) {
  rec <- fileTab[i,]
  eachTab <- read_delim(rec$filePath, delim = "\t", 
                        col_names = FALSE, col_types = "ci") %>%
    mutate(ID = rec$ID)
  }) %>% bind_rows() %>%
  pivot_wider(names_from = ID, values_from = X2)

countMat <- data.frame(dplyr::filter(countTab, !str_detect(X1,"__"))) %>%
  column_to_rownames("X1") %>% as.matrix()

unmappedTab <- dplyr::filter(countTab, str_detect(X1, "__")) %>%
  pivot_longer(-X1, names_to = "ID", values_to = "count") %>%
  mutate(X1 = str_remove_all(X1,"__")) %>%
  dplyr::rename(feature = X1)
```

```{r}
# Check all samples have annotations
all(fileTab$fileName %in% patAnno$fileName)
all(patAnno$fileName %in% fileTab$fileName)

# Modify rownames in patAnno and colnames in count matrix
countMat <- countMat[,colnames(countMat) %in% patAnno$ID]
countMat <- countMat[,patAnno$ID]  #order column names
```


# Assemble DEseq object

Create DESeq2 object called ddsDrug
```{r create-DESeq2-object}
ddsDrug <- DESeqDataSetFromMatrix(countMat, patAnno, design = ~1)
```

Annotate genes
```{r annotate-genes}
geneAnno <- transcripts(EnsDb.Hsapiens.v86,
                        columns = c("gene_id", "seq_name", "gene_name","gene_biotype"),
                        return.type = "DataFrame") %>% as_tibble() %>%
                        distinct(gene_id, .keep_all = TRUE)
geneAnno$tx_id <- NULL

# Annotate transcripts
rowAnno <- geneAnno %>% distinct(gene_id, seq_name, gene_name, gene_biotype) %>%
  dplyr::rename(chromosome = seq_name, symbol = gene_name, biotype = gene_biotype) %>%
  data.frame(stringsAsFactors = FALSE) %>% column_to_rownames("gene_id")

rowData(ddsDrug) <- rowAnno[rownames(ddsDrug),]
```

Post processing
```{r}
#remove genes never expressed
ddsDrug <- ddsDrug[rowSums(counts(ddsDrug)) !=0,]

#calculate size factor
ddsDrug <- estimateSizeFactors(ddsDrug)
```

# Summarise htseq statistics
```{r}
mappedTab <- tibble(ID = colnames(countMat),
                   feature = "total mapped",
                   count=colSums(countMat))

totalTab <- bind_rows(mappedTab, unmappedTab) %>%
  group_by(ID) %>% mutate(total = sum(count)) %>%
  ungroup() %>% mutate(percent = count/total) %>%
  mutate(fileName = fileTab[match(ID, fileTab$ID),]$fileName)

write_tsv(totalTab, "./output/htseq_process_batchPilot1c234_20220207.tsv")
```

# Save results
```{r}
save(ddsDrug, file = "./output/ddsDrug_htseq_batchPilot1c234_20220207.RData")
```

# Session Info Details
```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```
