---
title: "Quality assessment of trimmed drug perturbation RNAseq data of all batches, pre-processed with Salmon"
author: "Caroline Lohoff (based on code from Junyan Lu)"
date: "February 01 2022"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Description
The objective of the analysis is the quality control of the RNA sequencing data after perturbation with either DMSO or one of ten drugs. The RNAseq data from 1106 samples were pre-processed with Salmon with prior adapter trimming of 12 bases from 3'-end.
This analysis assesses and compares the quality of the five batches pilot, batch 1 combined, batch 2, batch 3, and batch 4.
Furthermore, the quality is compared to a previous RNAseq data set from 2019.

# Load data and packages

Set global options
```{r setup, include=FALSE, cache = FALSE}
#knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_knit$set(root.dir = "~/Documents/R/drugseq_test")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load packages
```{r}
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(tidyverse)
```

Load data
```{r}
load("./output/ddsDrug_trimmed_batch3_andPrevious_20211220.RData")
load("./data/patmeta_210324.RData")

# Add columns to DESeq object using patient meta data
ddsDrug$diagnosis <- patMeta[match(ddsDrug$patID, patMeta$Patient.ID),]$diagnosis
ddsDrug$methylation <- patMeta[match(ddsDrug$patID,
                                     patMeta$Patient.ID),]$Methylation_Cluster
ddsDrug$IGHVpat <- patMeta[match(ddsDrug$patID, patMeta$Patient.ID),]$IGHV.status
ddsDrug$IGHV <- ifelse(is.na(ddsDrug$IGHV), ddsDrug$IGHVpat, ddsDrug$IGHV)
ddsDrug$IGHVpat <- NULL
```

Set global ggplot variables
```{r}
theme_set(theme_bw() + theme(axis.text = element_text(size=12), 
                             axis.title = element_text(size=14),
                             plot.title = element_text(size = 15,
                                                       hjust =0.5,
                                                       face="bold")))
```

# Distribution of raw counts

## Before normalization

Define function for raw counts box plot
```{r}
raw_counts_plot <- function(inputTable){
  plot <- ggplot(inputTable, aes(x= ID, y= log10(count), fill = batch)) +
    geom_boxplot() + theme_bw() + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
  return(plot)
}
```

Plot data
```{r, fig.width=20, fig.height=5}
patAnno <- colData(ddsDrug) %>% as_tibble()
plotTab <- data.frame(counts(ddsDrug)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(batch = patAnno[match(ID, patAnno$ID),]$batch)

raw_counts_plot(plotTab)
```

## After normalization

All samples
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsDrug, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(batch = patAnno[match(ID, patAnno$ID),]$batch)

raw_counts_plot(plotTab)
```

Pilot batch and batch 3
```{r,fig.width=15, fig.height=5}
plotTab_batch3 <- dplyr::filter(plotTab, batch == c("pilot", "batch3"))

raw_counts_plot(plotTab_batch3)
```

# Check ECDFs

```{r}
# Remove rows/genes with too few counts
keep <- apply(counts(ddsDrug), 1, function(x) any(x >= 10))
ddsAll <- ddsDrug[keep,]
# Only use protein coding genes
ddsAll <- ddsAll[rowData(ddsAll)$biotype %in% "protein_coding",]
# Remove sex chromosome genes
ddsAll <- ddsAll[!rowData(ddsAll)$chromosome %in% c("X","Y")]
# Variance stabilization transformation of the raw data
RNAnorm.all <- vst(ddsAll)
```

```{r, fig.width=15, fig.height=7, cache=TRUE}
# Successful normalization should lead to overlapped curves
par(mfrow = c(1,2))
geneplotter::multidensity(assay(RNAnorm.all), xlim = c(0,30), legend = F, xlab = "mean counts")
geneplotter::multiecdf(assay(RNAnorm.all), legend = F, xlab="mean counts")
```


# Clustering -- PCA

## Consider all samples

### Calculate PCA
```{r, fig.height=6, fig.width=8}
#invariant filtering
exprMat <- assay(RNAnorm.all)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:5000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>%
  rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)
```

### PCA colored by batch
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = batch)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
The batches are separated PC2. Perhaps this is due to the trimming of the 3'end.

The numbers on the x-axis and y-axis describe the variance of the principle components.

### PCA colored by treatment
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = treatment)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
The treatments are not separated by the principal components. Only the baseline samples of batch 1, batch 2 and batch 3 are separated by PC2. 

## Remove the baseline samples of batches 1-3
We have seen that the baseline samples are separated from the other samples, as expected. To be able to see differences between the drug-perturbed samples, the nine baseline samples will be removed from the following plots showing the PCA results.
```{r}
remove <- dplyr::filter(pcaTab, treatment == "Baseline",
                        batch %in% c("batch1", "batch2", "batch3"))
pcaTab <- dplyr::anti_join(pcaTab, remove, by="ID")
```

### PCA colored by diagnosis
```{r}
ggplot(pcaTab, aes(x=PC1, y=PC2, col = diagnosis)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
Now we see that PC2 separates patients with a different diagnosis. CLL and mantle cell lymphoma (MCL) can be clearly separated.

### PCA colored by patientID
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = patID)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
As we have seen in the previous plot, the patients who were diagnosed with MCL (patient IDs P0644, P0683 and P0782) or with follicular lymphoma (FL) (patient ID P0057) are separated from the patients diagnosed with CLL.

## Focus on CLL

In this analysis we are only interested in CLL. Therefore, all samples from patients with a different diagnosis are being removed.
```{r}
pcaTab <- dplyr::filter(pcaTab, diagnosis == "CLL") # removing 97 samples
```

### PCA colored by batch
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = batch)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
The separation between batch 1 combined and the two batches 2 and 3 is even more visible in this plot. Since we want to use all the batches together for performing differential expression analysis and pathway enrichment analysis, the separation of the batches will introduce a bias. Thus, it is suggested to use the untrimmed RNAseq data preprocessed with Salmon.

### PCA colored by patientID
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = patID)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
The outliers caused by the samples of the patient with the ID P0066 will be removed from further analysis. This female patient was first diagnosed with CLL in 2006 and treated the first time in 2010. The immunoglobulin heavy chain gene (IgHV) is unmutated and thus the survival rate is worse than that of patients with mutated status. The overall level of programming per cluster is considered as low-programmed (LP-CLL).
```{r}
pcaTab <- dplyr::filter(pcaTab, !patID == "P0066") 
```

### PCA colored by treatment
```{r}
ggplot(pcaTab, aes(x=PC1,y=PC2, col = treatment)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) 
```
Since no difference can be determined on the basis of the treatment, it must be assumed that the viability is driven by apoptosis.

This could be a problem for further analysis, as we want to emphasize the different efficacy and modes of action of the drugs.

### PCA colored by IGHV status
```{r}
ggplot(pcaTab, aes(x=PC1, y=PC2, col = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
There is no separation based on IGHV status.

### PCA colored by average fragment size
```{r}
ggplot(pcaTab, aes(x=PC1, y=PC2, col = Average.fragment.size)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
There is no separation based on fragment sizes.

### PCA colored by library concentration
```{r}
ggplot(pcaTab, aes(x=PC1, y=PC2, col = Library.concentration)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
There is no separation based on library concentrations.

### PCA colored by methylation cluster
```{r}
ggplot(pcaTab, aes(x=PC1, y=PC2, col = methylation)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + ylab(sprintf("PC2 (%2.1f%%)",varExp[2]))
```
There is no separation based on methylation cluster, which can be split up into low-programmed (LP), intermediate-programmed (IP), and high-programmed (HP).

## Results shown in Heatmap

### Sample distance (only CLL)
```{r}
sampleDists <- dist(t(assay(RNAnorm.all)))
sampleDistMat <- as.matrix(sampleDists)
annoCol <- patAnno %>% select(expID, expBatch, patID, FSC.SSC.norm, treatment, time, batch, diagnosis, ID) %>%
  data.frame() %>% column_to_rownames("ID")

#keep only patient samples diagnosed with CLL, exclude outliers from patient P0066
annoColred <- dplyr::filter(annoCol, diagnosis == "CLL", !patID == "P0066") %>% 
  dplyr::select(-diagnosis)

sampleDistMatred <- sampleDistMat[rownames(sampleDistMat)%in%row.names(annoColred),
                           colnames(sampleDistMat)%in%row.names(annoColred)]

```

```{r, fig.width=20, fig.height=20}
#Plot heatmap
colors <- colorRampPalette( rev(brewer.pal(9,"Blues")) )(255)
pheatmap(sampleDistMatred, color = colors, clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, annotation_col = annoColred, clustering_method = "ward.D2",
         cluster_rows=F, cluster_cols=F,
         labels_col = paste0(patAnno[colnames(sampleDistMat),]$patID,"_",
                             patAnno[colnames(sampleDistMat),]$treatment))
```

### Sample similarity (only CLL)

Next, we plot the sample similarity which is based on Pearson correlation.
```{r, fig.width=20, fig.height=20}
sampleSimMat <- cor(assay(RNAnorm.all))
sampleSimMatred <- sampleSimMat[rownames(sampleSimMat)%in%row.names(annoColred),
                           colnames(sampleSimMat)%in%row.names(annoColred)]
pheatmap(sampleSimMatred, annotation_col = annoColred,
         clustering_method = "ward.D2",
         cluster_rows=F, cluster_cols=F,
         labels_col = paste0(patAnno[colnames(sampleSimMat),]$patID,"_",
                             patAnno[colnames(sampleSimMat),]$treatment))
```


# Compare with previous RNAseq data (samples with baseline included)

## Processing datasets
```{r}
load("~/Documents/R/drugseq_test/data/ddsCLL_Salmon_190702.RData")
ddsSub <- ddsDrug[,ddsDrug$treatment %in% "Baseline"]   # 9 samples in new RNA seq data contain baseline
overPat <- intersect(ddsCLL$patientID, ddsSub$patID)    # 5 overlapping patient IDs
ddsOld <- ddsCLL[,overPat]  
ddsNew <- ddsDrug[,match(overPat, ddsDrug$patID)]  

#only keep protein coding genes
ddsOld <- ddsOld[rowData(ddsOld)$gene_biotype %in% "protein_coding",]  # dim 22028 rows x 5 columns
ddsNew <- ddsNew[rowData(ddsNew)$biotype %in% "protein_coding",]       # dim 22028 rows x 5 columns
colnames(ddsNew) <- ddsNew$patID
```
5 overlapping patient IDs between old (2019) and new RNAseq (2021) data.

## Summarise detected genes
```{r, warning=FALSE, message=FALSE}
sumCountTab <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  geneOld <- rownames(subOld)[rowSums(assay(subOld)) > 0]
  geneNew <- rownames(subNew)[rowSums(assay(subNew)) > 0]
  tibble(patID = n,
         onlyNew = length(setdiff(geneNew, geneOld)),
         onlyOld = length(setdiff(geneOld, geneNew)),
         both = length(intersect(geneNew, geneOld)))
}) %>% bind_rows()
sumCountTab
```

```{r}
plotTab <- sumCountTab %>% pivot_longer(-patID, names_to = "group", values_to = "count") %>%
  mutate(group = factor(group, level = c("onlyNew","both","onlyOld")))
ggplot(plotTab, aes(x=patID, y = count, fill = group)) + geom_bar(stat = "identity") +
    theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
```

## Correlations of gene counts

```{r, fig.width=10, fig.height=10, message=FALSE, warning=FALSE}
plotList <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  overGene <- intersect(rownames(subOld), rownames(subNew))
  subOld <- subOld[overGene,]
  subNew <- subNew[overGene,]
  plotTab <- tibble(geneID = overGene,
                    countOld = assay(subOld)[,1],
                    countNew = assay(subNew)[,1])
  ggplot(plotTab, aes(x=log10(countOld), y=log10(countNew))) + scattermore::geom_scattermore() +
     ggtitle(n)
})
cowplot::plot_grid(plotList[[1]], plotList[[2]], plotList[[3]], plotList[[4]], plotList[[5]], ncol=3, labels="AUTO")
```


## Compare the similarity of expression pattern

Assemble new combined DESeq object
```{r}
oldMat <- counts(ddsOld)
newMat <- counts(ddsNew)
colnames(newMat) <- paste0(colnames(newMat),"_1")
patBack <- data.frame(row.names = c(colnames(oldMat), colnames(newMat)),
                      patID = c(ddsOld$patientID,ddsNew$patID),
                      set = rep(c("old","new"), each = ncol(oldMat)))
overGene <- intersect(rownames(oldMat), rownames(newMat))
comMat <- cbind(oldMat[overGene,],newMat[overGene,])
ddsCom <- DESeqDataSetFromMatrix(comMat, patBack, design= ~1)
ddsCom <- estimateSizeFactors(ddsCom)  #estimate size factor
```

Distribution after normalization
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsCom, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(set = patBack[ID,]$set)

ggplot(plotTab, aes(x= ID, y= log10(count))) +
  geom_boxplot(aes(fill = set)) + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

Variance stabilization transformation
```{r}
ddsCom.vst <- varianceStabilizingTransformation(ddsCom)
```

### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.vst)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

# create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```
The batch (or sequencing platform) explains the major difference

### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol, clustering_method = "ward.D2")
```


## Compare similarity of expression pattern after adjusting for batch

Adjust batch effect using limma
```{r}
exprMat <- assay(ddsCom.vst)
exprMat <- sva::ComBat(exprMat, batch = factor(ddsCom.vst$set))
ddsCom.adj <- ddsCom.vst
assay(ddsCom.adj) <- exprMat
```


### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom.adj) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.adj)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```


### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol, clustering_method = "ward.D2")
```


# Compare with previous RNAseq data (using DMSO as baseline)

## Processing datasets
```{r}
load("~/Documents/R/drugseq_test/data/ddsCLL_Salmon_190702.RData")
ddsSub <- ddsDrug[,ddsDrug$treatment %in% "DMSO"]    # 38 samples were treated with DMSO
overPat <- intersect(ddsCLL$patientID, ddsSub$patID) # 31 overlapping patient IDs
ddsOld <- ddsCLL[,overPat]  
ddsNew <- ddsDrug[,match(overPat, ddsDrug$patID)] 

#only keep protein coding genes
ddsOld <- ddsOld[rowData(ddsOld)$gene_biotype %in% "protein_coding",] # dim: 22028 rows x 31 columns
ddsNew <- ddsNew[rowData(ddsNew)$biotype %in% "protein_coding",]      # dim: 22028 rows x 31 columns
colnames(ddsNew) <- ddsNew$patID
```
31 patient IDs treated with DMSO are shared between old (2019) and new RNAseq (2021) data.


## Summarise detected genes
```{r}
sumCountTab <- lapply(overPat, function(n) {
  subOld <- ddsOld[,n]
  subNew <- ddsNew[,n]
  geneOld <- rownames(subOld)[rowSums(assay(subOld)) > 0]
  geneNew <- rownames(subNew)[rowSums(assay(subNew)) > 0]
  tibble(patID = n,
         onlyNew = length(setdiff(geneNew, geneOld)),
         onlyOld = length(setdiff(geneOld, geneNew)),
         both = length(intersect(geneNew, geneOld)))
}) %>% bind_rows()
sumCountTab
```

```{r}
plotTab <- sumCountTab %>% pivot_longer(-patID, names_to = "group", values_to = "count") %>%
  mutate(group = factor(group, level = c("onlyNew","both","onlyOld")))
ggplot(plotTab, aes(x=patID, y = count, fill = group)) + geom_bar(stat = "identity")
```
The overlap of the detected genes look good.
**both**: genes have non-zero count in both datasets
**onlyNew**: genes detected only in the current batch 1 new from 2021 (shallow sequencing)
**onlyOld**: genes detected only in the previous batch from 2019 (full depth)  


## Compare the similarity of expression pattern

Assemble new combined DESeq object
```{r}
oldMat <- counts(ddsOld)
newMat <- counts(ddsNew)
colnames(newMat) <- paste0(colnames(newMat),"_1")
patBack <- data.frame(row.names = c(colnames(oldMat), colnames(newMat)),
                      patID = c(ddsOld$patientID,ddsNew$patID),
                      set = rep(c("old","new"), each = ncol(oldMat)))
overGene <- intersect(rownames(oldMat), rownames(newMat))
comMat <- cbind(oldMat[overGene,],newMat[overGene,])
ddsCom <- DESeqDataSetFromMatrix(comMat, patBack, design= ~1)
ddsCom <- estimateSizeFactors(ddsCom)  #estimate size factor
```

Distribution after normalization
```{r,fig.width=15, fig.height=5}
plotTab <- data.frame(counts(ddsCom, normalized = TRUE)) %>% 
  rownames_to_column("id") %>%
  gather(key= "ID", value = "count",-id) %>%
  dplyr::filter(count > 0) %>%
  mutate(set = patBack[ID,]$set)

ggplot(plotTab, aes(x= ID, y= log10(count))) +
  geom_boxplot(aes(fill = set)) + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

Variance stabilization transformation
```{r}
ddsCom.vst <- varianceStabilizingTransformation(ddsCom)
```

### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.vst)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```
The batch (or sequencing platform) explains the major difference.

### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol,
          clustering_method = "ward.D2")
```


## Compare similarity of expression pattern after adjusting for batch

Adjust batch effect using limma
```{r}
exprMat <- assay(ddsCom.vst)
exprMat <- limma::removeBatchEffect(exprMat, batch = factor(ddsCom.vst$set))
ddsCom.adj <- ddsCom.vst
assay(ddsCom.adj) <- exprMat
```


### PCA

```{r, fig.height=6, fig.width=8, message=FALSE, warning=FALSE}
patAnno <- colData(ddsCom.adj) %>% data.frame() %>% rownames_to_column("ID") %>%
  mutate(IGHV = patMeta[match(patID, patMeta$Patient.ID),]$IGHV.status,
         trisomy12 = patMeta[match(patID, patMeta$Patient.ID),]$trisomy12)

#invariant filtering
exprMat <- assay(ddsCom.adj)
sds <- rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing  = T)[1:10000],]

#calculate pca 
pcaRes <- prcomp(t(exprMat), scale. = TRUE, center=TRUE)
varExp <- (pcaRes$sdev^2 / sum(pcaRes$sdev^2))*100
pcaTab <- data.frame(pcaRes$x[,1:10]) %>% rownames_to_column("ID") %>%
  left_join(patAnno)
names(varExp) <- colnames(pcaRes$x)

#create plot
ggplot(pcaTab, aes(x=PC1,y=PC2, col = set, shape = IGHV)) + geom_point() + theme_bw() + 
  xlab(sprintf("PC1 (%2.1f%%)",varExp[1])) + 
  ylab(sprintf("PC2 (%2.1f%%)",varExp[2])) +
  geom_line(aes(group = patID), col = "grey50", linetype = "dotted") +
  ggrepel::geom_text_repel(aes(label = patID))
```

### Sample similarity
```{r, fig.width=8, fig.height=8}
sampleSimMat <- cor(exprMat)
annoCol <- patAnno %>% select(ID, patID, set, IGHV) %>%
  column_to_rownames("ID")

pheatmap(sampleSimMat, annotation_col = annoCol,
          clustering_method = "ward.D2")
```


# Session Info Details
```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```
