---
title: "MOFA analysis"
author: "Junyan Lu"
date: "5/11/2022"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Description
In this script different normalization methods for count matrices are tested and used as input for MOFA. 

# Preparations
Set global options
```{r setup, include=FALSE, cache = FALSE}
knitr::opts_knit$set(root.dir = "~/Documents/R/drugseq_test")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load packages
```{r}
suppressPackageStartupMessages({
  library(MOFA2)
  library(DESeq2)
  library(limma)
  library(tidyverse)
  library(corrplot)
  library(pheatmap)
})
```

Lod data
```{r}
load("./data/patmeta_210324.RData")
load("./output/ddsDrug_htseq_batchPilot1c234_20220426.RData")
```

Subset conditions
```{r}
ddsDrug$IGHV <- patMeta[match(ddsDrug$patID,patMeta$Patient.ID),]$IGHV.status
ddsDrug$diagnosis <- patMeta[match(ddsDrug$patID,patMeta$Patient.ID),]$diagnosis

ddsDrug <- ddsDrug[,ddsDrug$batch!="pilot" & ddsDrug$diagnosis %in% "CLL" & 
                    #ddsDrug$treatment %in% c("DMSO","Nutlin-3a","Ibrutinib","Duvelisib") & 
                    ddsDrug$Replicate == 1]
ddsDrug <- ddsDrug[rowSums(counts(ddsDrug)) >0, ] #removed not expressed genes again because of subsetting

#annotated genes
#ddsDrug <- ddsDrug[rowData(ddsDrug)$biotype == "protein_coding" & !rowData(ddsDrug)$symbol %in% c(NA, ""),]
# Remove all mitochondrial and artificial chromosomes
```

# Abundance of different types of genes
```{r}
ddsDMSO <- ddsDrug[,ddsDrug$treatment=="DMSO"]
sumTab <- counts(ddsDMSO) %>% as_tibble(rownames = "id") %>%
  pivot_longer(-id) %>%
  mutate(patID = colData(ddsDMSO)[name,]$patID,
         treatment = colData(ddsDMSO)[name,]$treatment,
         biotype = rowData(ddsDMSO)[id,]$biotype,
         chr = rowData(ddsDMSO)[id,]$chromosome)
sumType <- group_by(sumTab, biotype, patID) %>%
  summarise(medVal = median(value), meanVal = mean(value))

ggplot(sumType, aes(x=biotype, y=log(medVal))) + geom_boxplot() + geom_point() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
ggplot(sumType, aes(x=biotype, y=log(meanVal))) + geom_boxplot() + geom_point() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
```
A lot of mitochondrial genes have a high abundance. In comparison the abundance of mitochondrial genes is higher than the abundance of genes from the chromosomes. One explanation, why mt_rRNA is so abundant could be that mitochondrial RNA is associated with viability. This could be hint that a lot of cells are dying, as the cellular mitochondrial content determines the apoptotic fate and modulates the time to death. Cells with higher mitochondrial content are more prone to die [see paper "Mitochondrial levels determine variability in cell death by modulating apoptotic gene expression" by MÃ¡rquez-Jurado et al. ](https://www.nature.com/articles/s41467-017-02787-4). In this paper they confer mitochondria a powerful discriminatory capacity of apoptotic fate. The results reveal a different role of mitochondria in apoptosis as the global regulator of apoptotic protein expression. 

Show number of genes per category
```{r, fig.height=6, fig.width=12}
geneType <- as.data.frame(table(sumTab$biotype))
ggplot(geneType, aes(x=Var1, y=Freq)) +
  geom_bar(stat="identity", fill="steelblue") +
  geom_text(aes(label=Freq), vjust=-0.3, size=2.5) +
  theme_minimal() +
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(angle=90, vjust=0.5, hjust=1))
```
For further analysis we only use protein_coding genes. 

Show log mean expression of genes per chromosome
```{r}
sumType <- group_by(sumTab, chr, patID) %>%
  summarise(medVal = median(value), meanVal = mean(value))

ggplot(sumType, aes(x=chr, y=log(medVal))) + geom_boxplot() + geom_point() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
ggplot(sumType, aes(x=chr, y=log(meanVal))) + geom_boxplot() + geom_point() +
  theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5))
```
The mean expression value of genes from chromosome X is very similar to genes on the chromosomes 1-22. Genes on chromosome Y show a different expression pattern. Hence, we only consider chromosomes 1-22 and X for further analysis. 

Select only chromosomes 1-22 and X
```{r}
ddsDrug <- ddsDrug[rowData(ddsDrug)$chromosome %in% c(1:22,"X","Y")]
```

# Correlation between size factors and viability
```{r}
ddsDrug <- estimateSizeFactors(ddsDrug)  # DESeq method
countMat <- counts(ddsDrug)
tCount <- colSums(countMat)
aCount <- colMeans(countMat)

countMat.norm <- counts(ddsDrug, normalize = TRUE)
tCount.norm <- colSums(countMat.norm)
aCount.norm <- colMeans(countMat.norm)

sizeTab <- tibble(smpID = colnames(ddsDrug),
                  sizeFactor = ddsDrug$sizeFactor,
                  viab = ddsDrug$FSC.SSC,
                  condition = ddsDrug$treatment,
                  viab0 = ddsDrug$FSC.SSC.d0,
                  viabDiff = viab0-viab,
                  totalCount = tCount,
                  totalCount.norm = tCount.norm,
                  avgCount = aCount,
                  avgCount.norm = aCount.norm) 

ddsDrug$viabDiff <- sizeTab$viabDiff
```

Correlation plot
```{r, fig.height=10, fig.width=10}
corTab <- sizeTab %>% dplyr::select(-condition) %>%
  data.frame() %>% column_to_rownames("smpID")
corrplot.mixed(cor(corTab, use = "pairwise.complete.obs"))
```
After library size correction using DESeq2, the size factors are correlated with the total counts and the average counts. 

Correlation of size factors and normalized average counts
```{r}
ggplot(sizeTab, aes(x=avgCount.norm, y=sizeFactor)) +
  geom_point() + geom_smooth()
```
There is also a correlation between size factors and normalized average counts. Thus, DESeq2's estimateSizeFactors function can not handle the sparse counts matrix and the large differences between the patients. The low sparsity in the count matrix may likely result from the shallow sequencing. Although shallow sequencing captures almost all genes that deep sequencing detects, the overall gene expression is much lower. In shallow sequening the vst normalied counts have values between 5 and 11 and in deep sequening around 20. The sparsity of the count matrix makes our shallow sequencing data set more like a single cell data set. The strong inter-patient heterogeneity can be compared with different cell types and the samples per patients would be the single cells. 

In order to deal with the low sparsity and the strong inter-patient heterogeneity, another correction method needs to be applied. In the next chapter two normalization methods are compared two each other. On the one hand we test the deconvolution method from Marioni lab and on the other hand the scTransform method from Satija lab. Both methods calculate size factors and normalize the counts in the gene expression matrix. The resulting counts matrix contains integers and can be used as input for DESeq2.

# Using deconvolution for normalization
[Marioni Lab, method Deconvolution 2016](https://github.com/MarioniLab/Deconvolution2016).

Cell-specific biases are normalized using the computeSumFactors method, which implements the deconvolution strategy for scaling normalization (Lun, Bach, and Marioni 2016).

```{r}
library(scran)

# Create new ddsAll object
ddsAll.deconv <- ddsDrug

# Reset size factors so that they can be calculated again
# (excluding this step has no consequences)
ddsAll.deconv$sizeFactor <- 1

# Determine clusters and use them to calculate size factors
clusters <- scran::quickCluster(ddsAll.deconv)
ddsAll.deconv <- scran::computeSumFactors(ddsAll.deconv,
                                   clusters=clusters)
# Show summary of size factors
summary(sizeFactors(ddsAll.deconv))

# Use size factors to normalize the counts
ddsAll.deconv <- scuttle::logNormCounts(ddsAll.deconv)
countMat.deconv <- counts(ddsAll.deconv)
```

Plot standard deviation and rank mean on deconvoluted counts
```{r}
vsn::meanSdPlot(countMat.deconv)
```
Advantage is the same as for scTransform. The resulting counts matrix contains integers and thus can be used as input for DESeq2. 
The integers are a bit higher than the ones receibed with scTransform.

## Variance modelling
This analysis is based on the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/scran/inst/doc/scran.html#1_Introduction).

We identify genes that drive biological heterogeneity in the data set by modelling the per-gene variance. By only using a subset of highly variable genes in downstream analyses like clustering, we improve resolution of biological structure by removing uninteresting genes driven by technical noise. We decompose the total variance of each gene into its biological and technical components by fitting a trend to the endogenous variances (Lun, McCarthy, and Marioni 2016). The fitted value of the trend is used as an estimate of the technical component, and we subtract the fitted value from the total variance to obtain the biological component for each gene.
```{r}
dec <- scran::modelGeneVar(ddsAll.deconv)
plot(dec$mean, dec$total, xlab="Mean log-expression", ylab="Variance")
curve(metadata(dec)$trend(x), col="blue", add=TRUE) 
```

If we have some uninteresting factors of variation, we can block these. This will perform the trend fitting and decomposition within each block before combining the statistics across blocks for output. Statistics for each individual block can also be extracted for further inspection. Modelling the per-gene variance is still very fast (< 5 Min). Blocking the patients leads to less variance. 
```{r}
# Blocking for patient ID to remove these factor of variation. 
# Additionally, we turn off weighting to avoid over-fitting for each patient
dec2 <- scran::modelGeneVar(ddsAll.deconv,
                     block=ddsAll.deconv$patID,
                     density.weights=FALSE)

plot(dec2$mean, dec2$total,
     xlab="Mean log-expression", ylab="Variance")

per.block <- dec2$per.block
par(mfrow=c(3, 2))
#for (i in seq_along(per.block)) {
#    decX <- per.block[[i]]
#    plot(decX$mean, decX$total, xlab="Mean log-expression", 
#        ylab="Variance", main=names(per.block)[i])
#    curve(metadata(decX)$trend(x), col="blue", add=TRUE)
#}
```

Extract some top genes for use in downstream procedures using the getTopHVGs() function.
A variety of different strategies can be used to define a subset of interesting genes:
```{r}
# Get the top 10% of genes
top.hvgs <- scran::getTopHVGs(dec, prop=0.1)
head(top.hvgs, n=10)

# Get the top 2000 genes
top.hvgs2 <- scran::getTopHVGs(dec, n=2000)
head(top.hvgs2, n=10)

# Get all genes with positive biological components
top.hvgs3 <- scran::getTopHVGs(dec, var.threshold=0)
head(top.hvgs3, n=10)

# Get all genes with FDR below 5%
top.hvgs4 <- scran::getTopHVGs(dec, fdr.threshold=0.05)
head(top.hvgs4, n=10)
```


# Using scTransform for normalization

[scTransform paper](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1), [GitHub](https://github.com/satijalab/sctransform)

```{r, results=FALSE}
library(sctransform)

countMat <- counts(ddsDrug)

# Correct count matrix with scTransform
scModel <- sctransform::vst(countMat,
                            return_corrected_umi = TRUE)
```

Extract normalized counts matrix
(Advantage: counts are integers and can be used as input for DESeq2)
```{r}
countMat.norm <- as.matrix(scModel$umi_corrected)

# Create new ddsAll object with norm. counts matrix and colData from ddsDrug
ddsAll <- DESeqDataSetFromMatrix(countMat.norm,
                                 colData = colData(ddsDrug), design = ~1)
rowData(ddsAll) <- rowData(ddsDrug[rownames(countMat.norm)],)
## size factors are the same as in ddsDrug
```

Extract vst matrix
(floats around 0, also negative counts)
```{r}
countMat.sc <- as.matrix(scModel$y)

# Create new ddsAll object with vst count matrix
ddsAll.sc <- ddsAll
assay(ddsAll.sc) <- countMat.sc
## size factors are the same as in ddsDrug
```

Plot standard deviation and rank mean on normalized counts and vst counts
```{r}
vsn::meanSdPlot(countMat.norm)
vsn::meanSdPlot(countMat.sc)
```

Compare norm. count matrix with sc count matrix
```{r}
# Create plot table (gene, sample ID, count)
plotTab.count <- countMat.norm[,ddsAll$treatment=="DMSO"] %>% 
  as_tibble(rownames = "id") %>%
  pivot_longer(-id)

plotTab.vst <- countMat.sc[,ddsAll$treatment=="DMSO"] %>%
  as_tibble(rownames = "id") %>%
  pivot_longer(-id)

ggplot(plotTab.count, aes(x=name, y = log2(value))) +
  geom_boxplot() + ggtitle("Raw counts normalized")
ggplot(plotTab.vst, aes(x=name, y=value)) +
  geom_boxplot() + ggtitle("Counts normalized by size factors (vst)")
```

## Use DESeq2::vst and limma::voom on unnormalized counts

```{r}
# Use vst function from DESeq2
ddsAll.vst <- ddsDrug
ddsAll.vst <- DESeq2::vst(ddsAll.vst)
countMat.vst <- assay(ddsAll.vst)

# Use voom function from limma
countMat <- counts(ddsDrug)
countMat.voom <- limma::voom(countMat,
                             lib.size = ddsAll$sizeFactor)$E
```

Plot standard deviation and rank mean for vst matrix
```{r}
vsn::meanSdPlot(countMat.vst)
```

Plot standard deviation and rank mean for voom matrix
```{r}
vsn::meanSdPlot(countMat.voom)
```

Correlation plot
```{r}
plot(countMat.vst[5,], countMat.voom[5,])
```

## Use DESeq2::vst and limma::voom on counts normalized with scTransform

VST transformation function from DESeq2 as comparison
```{r}
# Create new ddsAll object from norm. counts
ddsAll.vst <- ddsAll

# Reset size factors so that DESeq can calculate them again
ddsAll.vst$sizeFactor <- 1
#ddsSub <- ddsAll[rowMedians(counts(ddsAll))>0,]

# Use vst function from DESeq2
ddsAll.vst <- DESeq2::vst(ddsAll.vst)
countMat.vst <- assay(ddsAll.vst)
## size factors are still 1?!
## counts are floats between 5 and 11
```

Plot standard deviation and rank mean on vst counts
```{r}
vsn::meanSdPlot(countMat.vst)
```

Limma Voom transformation for comparison
```{r}
countMat <- counts(ddsAll)
countMat.voom <- limma::voom(countMat,
                             lib.size = ddsAll$sizeFactor)$E
```

Plot standard deviation and rank mean on voom transformed counts
```{r}
vsn::meanSdPlot(countMat.voom)
```

Correlation plot
```{r, fig.height=10, fig.width=10}
sizeTab.sub <- sizeTab[match(colnames(countMat.norm), sizeTab$smpID),]
sizeTab.sub$avgCount.sc <- colMeans(countMat.norm)
sizeTab.sub$avgVst.sc <- colMeans(countMat.sc)
sizeTab.sub$avgVst.vst <- colMeans(countMat.vst)
sizeTab.sub$avgVst.voom <- colMeans(countMat.voom)

corTab <- sizeTab.sub %>% dplyr::select(-condition) %>%
  data.frame() %>% column_to_rownames("smpID")
corrplot.mixed(cor(corTab, use = "pairwise.complete.obs"))
```

```{r}
plot(countMat.vst[5,], countMat.voom[5,])
plot(countMat.sc[5,], countMat.voom[5,])
```
Linear relationship between vst normalized counts and voom normalized counts, but relationship between sc normalized counts and voom is hyperbolic. This means that Pearson residuals (result of scTransform) are not normally distributed.

Histogram of counts normalized with scTransform shows Poisson distribution
```{r}
hist(rowMeans(assay(ddsAll)), breaks = 100)
```

Histogram of Pearson residuals obtained with scTransform
```{r}
hist(rowMeans(assay(ddsAll.sc)), breaks = 100)
```


SC transformation with also adjusting for patient effect
```{r, eval=FALSE}
## Takes minimum of 40 minutes!
countMat <- counts(ddsDrug)
patTab <- data.frame(colData(ddsDrug))[,c("patID","ID"),drop=FALSE]
scModel <- sctransform::vst(countMat,
                            cell_attr = patTab,
                            return_corrected_umi = TRUE,
                            latent_var_nonreg  = "patID")
save(scModel, file = "./output/scModel_adjusted.RData")
countMat.norm <- as.matrix(scModel$umi_corrected)
countMat.sc <- as.matrix(scModel$y)

ddsAll.adj <- DESeqDataSetFromMatrix(countMat.norm, colData = colData(ddsDrug), design = ~1)
rowData(ddsAll.adj) <- rowData(ddsDrug[rownames(countMat.norm)],)
ddsAll.adj.sc <- ddsAll.adj
assay(ddsAll.adj.sc) <- countMat.sc
```

```{r,eval=FALSE}
vsn::meanSdPlot(countMat.sc)
```


# Preprocessing using limma

Subset genes (vst normalized with scTransform)
```{r}
#annotated genes
ddsSub <- ddsAll.sc[rowData(ddsAll.sc)$biotype == "protein_coding" & !rowData(ddsAll.sc)$symbol %in% c(NA, ""),]
#ddsSub <- ddsAll.vst[rowData(ddsAll.vst)$biotype == "protein_coding" & !rowData(ddsAll.vst)$symbol %in% c(NA, ""),]
# Remove all mitochondrial and artificial chromosomes
#ddsSub <- ddsSub[rowData(ddsSub)$chromosome %in% c(1:22,"X","Y")]

# Remove genes on Y chromosome which could introduce some bias
ddsSub <- ddsSub[rowData(ddsSub)$chromosome != "Y"]
# ddsAll <- ddsAll[!rowData(ddsAll)$chromosome %in% c("X","Y")]

dim(ddsSub)
```

Order matrix by sds
```{r}
exprMat <- assay(ddsSub)
sds <- genefilter::rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing = TRUE),]

plotTab <- tibble(gene1=exprMat[1,], gene2= exprMat[3,], smpID = colnames(exprMat),
                  treatment = ddsSub$treatment,
                  patID = ddsSub$patID)

ggplot(plotTab, aes(x=gene1, y=gene2)) +
  geom_point(aes(col=patID)) + theme(legend.position = "none")
```
If we compare the expression of two genes, there are many outlier samples. Samples with a low mapping rate have some genes with a high expression. By using sds these genes get over-normalized. 

Order matrix by mad (median absolute deviation)
```{r}
exprMat <- assay(ddsSub)
sds <- apply(exprMat, 1, function(x) mad(x))
exprMat <- exprMat[order(sds, decreasing = TRUE),]

plotTab <- tibble(gene1=exprMat[1,], gene2= exprMat[3,], smpID = colnames(exprMat),
                  treatment = ddsSub$treatment,
                  patID = ddsSub$patID)

ggplot(plotTab, aes(x=gene1, y=gene2)) +
  geom_point(aes(col=patID)) + theme(legend.position = "none")
```
***Using mad for ordering the genes improves the distribution of samples.***

## Adjust for patient effect

Instead of adjusting for patient effect withing the formula of sc transform, we can simply use ComBat. This is much faster and has a similar effect. The removeBatchEffect function from the limma package has a similar effect, but ComBat is preferred here, since ComBat adjusts for differences in both the mean and variance differences across the batches, whereas limma assumes that the batch variances are the same and only accounts for mean differences across the batches. As there are large differences between the patients, ComBat is the method of choice.
```{r}
ddsVoom <- ddsSub
exprMat <- assay(ddsVoom)
exprMat <- sva::ComBat(exprMat, batch = factor(ddsVoom$patID))
#exprMat <- limma::removeBatchEffect(exprMat, batch = factor(ddsVoom$patID))
ddsCombat <- ddsSub
assay(ddsCombat) <- exprMat
```

```{r}
vsn::meanSdPlot(assay(ddsCombat))
hist(rowMeans(assay(ddsCombat)), breaks = 100)
```

Adjust for patient effect (use the adjusted scModel)
```{r, eval=FALSE}
ddsAll.adj.sc.sub <- ddsAll.adj.sc[rownames(ddsSub),]
ddsCombat <- ddsAll.adj.sc.sub
```

Subset for selected drugs
```{r}
drugs <- c("DMSO","Nutlin-3a","Ibrutinib","Duvelisib")
ddsCombat.sub <- ddsCombat[,ddsCombat$treatment %in% drugs] 
ddsVoom.sub <- ddsVoom[,ddsVoom$treatment %in% drugs] 
```


# PCA on scTransformed vst counts adjusted by patID using ComBat

Prefiltering
```{r}
exprMat<-assay(ddsCombat.sub)
# Using mad instead of rowSds
sds <- apply(exprMat,1,function(x) mad(x))
meanExpr <- colMeans(exprMat)
#sds <- genefilter::rowSds(exprMat)
exprMat <- exprMat[order(sds, decreasing = TRUE)[1:5000],]
```

Calculate PCA
```{r}
pcRes <- prcomp(t(exprMat), center = T, scale. = T)
pcTab <- pcRes$x[,1:100] %>%
    as_tibble(rownames = "smpID") %>%
    left_join(as_tibble(colData(ddsCombat.sub),
                        rownames = "smpID"), by = "smpID") %>%
    mutate(meanExp = meanExpr)
```

Variance explained by top 10 PC
```{r}
library(factoextra)
fviz_screeplot(pcRes, addlabels=TRUE, ylim = c(0, 9))
```
PC1 explains now 7.6% instead of 16.7%. Thus, transforming the counts using sc transform and correcting for patient ID has a strong impact. 

Color PCA by viability
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = FSC.SSC)) +
    geom_point()
```
Viability is not PC1 anymore. 

Color PCA by IGHV
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = IGHV)) +
    geom_point()
```
IGHV is not described by PC1 or PC2. Thus, adjusting for patient effect has worked out. 

Color PCA by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = treatment)) +
    geom_point()
```
PC1 is the drug response, as DMSO samples cluster. This is better visible if we plot PC1 vs PC3.

Color PCA by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC3, col = treatment)) +
    geom_point()
```

Color PCA by treatments
```{r}
ggplot(pcTab, aes(x=PC6, y=PC7, col = treatment)) +
    geom_point()
```

PC1 vs Mean Expression
```{r}
ggplot(pcTab, aes(x=PC1, y=meanExp)) + geom_point()
```

PC1 vs viability
```{r}
ggplot(pcTab, aes(x=PC1, y=FSC.SSC)) + geom_point()
```

Correlation between PCs and treatment types
```{r}
testTab <- dplyr::select(pcTab,smpID, contains("PC"), treatment) %>%
    pivot_longer(-c("smpID","treatment"), names_to = "PC", values_to = "value")

resTab <- group_by(testTab, PC) %>% nest() %>%
    mutate(m=map(data, ~anova(lm(value~treatment,.)))) %>%
    mutate(res = map(m, broom::tidy)) %>%
    unnest(res) %>% arrange(p.value) %>%
    dplyr::select(PC, p.value)

resTab
ggplot(pcTab, aes(x=treatment, y=PC1)) +
    geom_point() + geom_boxplot()
```

## UMAP
```{r, cache=FALSE}
library(smallvis)
#Calculate UMAP layout, which can better retain global structure
plotTab <- smallvis(t(exprMat), method = "umap", perplexity = 25, 
                    eta = 0.01, epoch_callback = FALSE, verbose = FALSE)
colnames(plotTab) <- c("umap1","umap2")

plotTab <- plotTab %>% as_tibble() %>% mutate(smpID = colnames(exprMat)) %>%
  left_join(as_tibble(colData(ddsCombat.sub), rownames = "smpID"), by = "smpID") 

ggplot(plotTab, aes(x=umap1, y=umap2, col = treatment)) +
    geom_point()
```


# PCA on counts without patient effect adjustment

Prefiltering
```{r}
exprMat<-assay(ddsVoom.sub)
meanExpr <- colMeans(exprMat)
# using mad instead of rowSds
sds <- apply(exprMat, 1, function(x) mad(x))
exprMat <- exprMat[order(sds, decreasing = TRUE)[1:5000],]
```

Calculate PCA
```{r}
pcRes <- prcomp(t(exprMat), center = T, scale. = T)
pcTab <- pcRes$x[,1:100] %>%
    as_tibble(rownames = "smpID") %>%
    left_join(as_tibble(colData(ddsVoom.sub),
                        rownames = "smpID"), by = "smpID") %>%
  mutate(meanExp = meanExpr)
```

Variance explained by top 10 PC
```{r}
library(factoextra)
fviz_screeplot(pcRes, addlabels=TRUE, ylim = c(0, 13))
```

PCA colored by viability
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = FSC.SSC)) +
    geom_point()
```
PC1 is viability

PCA colored by IGHV
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = IGHV)) +
    geom_point()
```

PCA colored by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC2, col = treatment)) +
    geom_point()
```

PCA colored by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC3, col = treatment)) +
    geom_point()
```

PCA colored by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC4, col = treatment)) +
    geom_point()
```

PCA colored by treatments
```{r}
ggplot(pcTab, aes(x=PC1, y=PC5, col = treatment)) +
    geom_point()
```

PCA colored by treatments
```{r}
ggplot(pcTab, aes(x=PC5, y=PC7, col = treatment)) +
    geom_point()
```

PC1 vs Mean expression (add correlation coefficient)
```{r}
ggplot(pcTab, aes(x=PC1, y=meanExp)) + geom_point()
```

Correlation PC1 and viability
```{r}
ggplot(pcTab, aes(x=PC1, y=FSC.SSC)) + geom_point()
```

Correlation between PCs and treatment types
```{r}
testTab <- dplyr::select(pcTab,smpID, contains("PC"), treatment) %>%
    pivot_longer(-c("smpID","treatment"), names_to = "PC", values_to = "value")

resTab <- group_by(testTab, PC) %>% nest() %>%
    mutate(m=map(data, ~anova(lm(value~treatment,.)))) %>%
    mutate(res = map(m, broom::tidy)) %>%
    unnest(res) %>% arrange(p.value) %>%
    dplyr::select(PC, p.value)

resTab
ggplot(pcTab, aes(x=treatment, y=PC7)) +
    geom_point() + geom_boxplot()
```


# MOFA on patient adjusted matrix

As input for MOFA we use the vst normalized counts by scTransform, which were additionally adjusted for patient effect.
```{r}
ddsMofa <- ddsCombat.sub
#sds <- genefilter::rowSds(assay(ddsMofa))
sds <- apply(assay(ddsMofa),1,function(x) mad(x))
ddsMofa <- ddsMofa[order(sds, decreasing = TRUE)[1:5000],]

exprObj <- lapply(unique(ddsMofa$treatment), function(x) {
  ddsEach <- ddsMofa[,ddsMofa$treatment==x]
  exprMat <- assay(ddsEach)
  colnames(exprMat) <- ddsEach$patID
  exprMat
})
names(exprObj) <- unique(ddsMofa$treatment)
```

Prepare genomics
```{r}
geneTab <- dplyr::filter(patMeta, Patient.ID %in% unique(ddsCombat.sub$patID)) %>%
    dplyr::select(Patient.ID, IGHV.status:U1) %>%
    #select(-Methylation_Cluster) %>%
    mutate(across(!contains("Patient"), as.factor)) %>%
    mutate(across(!contains("Patient"), as.numeric)) %>%
    column_to_rownames("Patient.ID") %>% as.matrix()
geneTab <- geneTab-1

# Use only genes with mutation in at least 5 patients
geneTab <- geneTab[,colSums(!is.na(geneTab))/nrow(geneTab) > 0.6 &
                       colSums(geneTab == 1, na.rm = TRUE) >= 5]
geneMat <- t(geneTab)

exprObj$Gene = geneMat
```

Define function for running MOFA
```{r}
running_mofa <- function(obj, var=0.01, mode="slow",
                         nFac=20, maxIter=10000){
  
  ### Setup MOFA training parameters
  # Define data options
  DataOptions <- get_default_data_options(obj)

  #Define model options
  ModelOptions <- get_default_model_options(obj)
  #ModelOptions$likelihoods["Gene"] <- "bernoulli"
  # number of factors to start with
  # (in each iteration number of factors is tried to be reduced)
  ModelOptions$num_factors <- nFac
  #ModelOptions

  # Define training options
  TrainOptions <- get_default_training_options(obj)
  TrainOptions$convergence_mode <- mode
  TrainOptions$maxiter <- maxIter
  # Automatically drop factors that explain less than x% of variance
  TrainOptions$drop_factor_threshold <- var
  #TrainOptions
  
  ### Run MOFA model
  obj_prep <- prepare_mofa(
    obj, 
    data_options = DataOptions,
    model_options = ModelOptions,
    training_options = TrainOptions
  )

  obj_res <- run_mofa(obj_prep)
  return(obj_res)
}
```

```{r}
# Create object
mofaData <- exprObj

# Create MultiAssayExperiment object 
mofaData <- MultiAssayExperiment::MultiAssayExperiment(
  experiments = mofaData)
```

Only keep samples that have at all five views
```{r, eval=FALSE}
useSamples <- MultiAssayExperiment::sampleMap(mofaData) %>%
  as_tibble() %>% group_by(primary) %>% summarise(n= length(assay)) %>%
  filter(n >= 5) %>% pull(primary)
mofaData <- mofaData[,useSamples]
```

```{r}
MOFAobject <- create_mofa_from_MultiAssayExperiment(mofaData)
MOFAobject
plot_data_overview(MOFAobject)
```
MOFA only uses samples that occur in all five views. 

```{r, eval=FALSE}
MOFAobject_res <- running_mofa(obj=MOFAobject, var=0.01, mode="fast")
save(MOFAobject_res, file = "./output/MOFAres_adjustedPat.RData")
## 86 iterations (ELBO=-2371021.66, deltaELBO=100.871 (0.00038363%), Factors=14)
```

Load results and show factors per view
```{r}
load("./output/MOFAres_adjustedPat.RData")
plot_variance_explained(MOFAobject_res, x="view", y="factor")
```

Factor correlation matrix
```{r, fig.width=5, fig.height=5}
plot_factor_cor(MOFAobject_res)
```
No correlation bewteen the factors. 

Total variance explained
```{r, fig.width=5, fig.height=4}
plot_variance_explained(MOFAobject_res, plot_total = T)[[2]]
```

## Factor heatmap

Contributions of patients to factors
```{r, fig.height=6, fig.width=15}
#gene annotation
facMat <- t(get_factors(MOFAobject_res)[[1]])

colAnno <- tibble(Name = colnames(facMat)) %>%
    mutate(IGHV = geneTab[Name, "IGHV.status"],
           trisomy12 = geneTab[Name, "trisomy12"],
           TP53 = geneTab[Name, "TP53"]) %>%
    data.frame() %>% column_to_rownames("Name")

pheatmap(facMat, clustering_method = "complete",
         annotation_col = colAnno,
         scale = "none")
```

Column medians vs factors of Nutlin
```{r}
fac <- facMat[2,]
facNut <- fac[names(fac) %in% colnames(exprObj$`Nutlin-3a`)]
medCol <- colMedians(exprObj$`Nutlin-3a`[,names(facNut)])
plot(facNut, medCol)
```

Weight of genomic features on LF1
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 1,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

Weight of genomic features on LF6
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 6,
 nfeatures = 10,
 scale = T
)
```

Weight of genomic features on LF7
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 7,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

Weight of genomic features on LF8
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 8,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```


# MOFA on un-adjusted matrix

As input for MOFA we use the count matrix adjusted with scTransform (vst), but not additionally adjusted for patient effect. 
```{r}
ddsMofa <- ddsVoom.sub
sds <- apply(assay(ddsMofa), 1, function(x) mad(x))
ddsMofa <- ddsMofa[order(sds, decreasing = TRUE)[1:5000],]

exprObj <- lapply(unique(ddsMofa$treatment), function(x) {
  ddsEach <- ddsMofa[,ddsMofa$treatment==x]
  exprMat <- assay(ddsEach)
  colnames(exprMat) <- ddsEach$patID
  exprMat
})
names(exprObj) <- unique(ddsMofa$treatment)
```

Prepare genomics
```{r}
geneTab <- filter(patMeta, Patient.ID %in% unique(ddsCombat.sub$patID)) %>%
    select(Patient.ID, IGHV.status:U1) %>%
    mutate(across(!contains("Patient"), as.factor)) %>%
    mutate(across(!contains("Patient"), as.numeric)) %>%
    column_to_rownames("Patient.ID") %>% as.matrix()
geneTab <- geneTab-1

geneTab <- geneTab[,colSums(!is.na(geneTab))/nrow(geneTab) > 0.6 &
                       colSums(geneTab == 1, na.rm = TRUE) >= 3]
geneMat <- t(geneTab)

exprObj$Gene = geneMat
```

Create MOFA object
```{r}
mofaData <- exprObj

# Create MultiAssayExperiment object 
mofaData <- MultiAssayExperiment::MultiAssayExperiment(
  experiments = mofaData)
```

Only keep samples that have at all five view
```{r, eval=FALSE}
useSamples <- MultiAssayExperiment::sampleMap(mofaData) %>%
  as_tibble() %>% group_by(primary) %>% summarise(n= length(assay)) %>%
  filter(n >= 5) %>% pull(primary)
mofaData <- mofaData[,useSamples]
```

```{r}
MOFAobject <- create_mofa_from_MultiAssayExperiment(mofaData)
plot_data_overview(MOFAobject)
```

```{r, eval=FALSE}
MOFAobject_res <- running_mofa(obj=MOFAobject, var=0.01, mode ="fast")
save(MOFAobject_res, file = "./output/MOFAres_unadjustedPat.RData")
## 136 iterations (ELBO=-3026034.40, deltaELBO=109.468 (0.00040025%), Factors=16)
```

Load result and show factors per view
```{r}
load("./output/MOFAres_unadjustedPat.RData")
plot_variance_explained(MOFAobject_res, x="view", y="factor")
```
Without adjusting for patients, more factors are shared across treatments. Thus, adjusting for patient effect leads to more distinct results.

Factor correlation matrix
```{r, fig.width=5, fig.height=5}
plot_factor_cor(MOFAobject_res)
```

Total variance explained
```{r, fig.width=5, fig.height=4}
plot_variance_explained(MOFAobject_res, plot_total = T)[[2]]
```

## Factor heatmap

```{r, fig.height=6, fig.width=15}
#gene annotation
facMat <- t(get_factors(MOFAobject_res)[[1]])[1:10,]

colAnno <- tibble(Name = colnames(facMat)) %>%
    mutate(IGHV = geneTab[Name, "IGHV.status"],
           trisomy12 = geneTab[Name, "trisomy12"],
           TP53 = geneTab[Name, "TP53"],
           KRAS = geneTab[Name, "KRAS"]) %>%
    data.frame() %>% column_to_rownames("Name")

pheatmap(facMat, clustering_method = "complete",
         annotation_col = colAnno, scale = "none")
```

Column medians vs factors of Nutlin
```{r}
fac <- facMat[2,]
# Vectors need to have the same length
facNut <- fac[names(fac) %in% colnames(exprObj$`Nutlin-3a`)]
# Order Nutlin columns by factor names
medCol <- colMedians(exprObj$`Nutlin-3a`[,names(facNut)])
plot(facNut, medCol)
```

Weight of genomic features on LF1
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 1,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```
Factor 1 is again IGHV status and methylation cluster

Weight of genomic features on LF2
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 2,
 nfeatures = 10,
 scale = T 
)
```
Interestingly, factor 2 is KRAS.

Weight of genomic features on LF3
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 3,
 nfeatures = 10,   
 scale = T      
)
```

Weight of genomic features on LF4
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 4,
 nfeatures = 10, 
 scale = T 
)
```
As factors 4 is mainly explained by the variance in Ibrutinib, patients with trisomy 19 are most affected by Ibrutinib treatment in comparison to the other drugs. Trisomy 19 is often correlated with trisomy 12. 

Weight of genomic features on LF6
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 6,
 nfeatures = 10,  
 scale = T     
)
```

Weight of genomic features on LF7
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 7,
 nfeatures = 10,  
 scale = T     
)
```
Factor 7 is also trisomy 19 and trisomy 12, but this time explained by the variance in Duvelisib. Thus, Ibrutinib and Duvelisib have a strong effect on patients with trisomy 19.
This result suggests an interaction between trisomy 12/19 and BCR signalling. Presumably, trisomy 12/19 enhances BCR signalling. 
***Interesting analysis: Conduct DESeq with design formula: trisomy12 + trisomy12:patIDgroup + trisomy12:treatment for Ibrutinib and Duvelisib. It is expected that patients with trisomy 12 (and trisomy 19) have more DE genes than other patients.***

Weight of genomic features on LF8
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 8,
 nfeatures = 10,  
 scale = T     
)
```


# MOFA on un-adjusted matrix minus DMSO

As input for MOFA we use the count matrix adjusted with scTransform (vst), but not additionally adjusted for patient effect. 
We subset this count matrix into the different treatments. The DMSO counts matrix is subtracted from each count matrix containing the counts for only one treatment. 

```{r}
usePat <- table(ddsVoom.sub$patID)
usePat <- names(usePat[usePat==4])
ddsMofa <- ddsVoom.sub[,ddsVoom.sub$patID %in% usePat]

ddsDMSO <- ddsMofa[,ddsMofa$treatment == "DMSO"]
ddsDrug <- ddsMofa[,ddsMofa$treatment != "DMSO"]
exprDMSO <- assay(ddsDMSO)

exprObj <- lapply(unique(ddsDrug$treatment), function(x) {
  ddsEach <- ddsMofa[,ddsMofa$treatment==x]
  exprMat <- assay(ddsEach)
  
  # Subtract DMSO matrix from each treatment matrix
  exprMat <- exprMat-exprDMSO
  
  # Only keep 5000 most variant genes 
  sds <- apply(exprMat, 1, function(x) mad(x))
  exprMat<- exprMat[order(sds, decreasing = TRUE)[1:5000],]

  colnames(exprMat) <- ddsEach$patID
  exprMat
})
names(exprObj) <- unique(ddsDrug$treatment)
```

Prepare genomics
```{r}
geneTab <- filter(patMeta, Patient.ID %in% unique(ddsCombat.sub$patID)) %>%
    select(Patient.ID, IGHV.status:U1) %>%
    mutate(across(!contains("Patient"), as.factor)) %>%
    mutate(across(!contains("Patient"), as.numeric)) %>%
    column_to_rownames("Patient.ID") %>% as.matrix()
geneTab <- geneTab-1

geneTab <- geneTab[,colSums(!is.na(geneTab))/nrow(geneTab) > 0.6 &
                       colSums(geneTab == 1, na.rm = TRUE) >= 3]
geneMat <- t(geneTab)

exprObj$Gene = geneMat
```

```{r}
# Create object
mofaData <- exprObj

# Create MultiAssayExperiment object 
mofaData <- MultiAssayExperiment::MultiAssayExperiment(
  experiments = mofaData)
```

Only keep samples that have at all five view
```{r, eval=FALSE}
useSamples <- MultiAssayExperiment::sampleMap(mofaData) %>%
  as_tibble() %>% group_by(primary) %>% summarise(n= length(assay)) %>%
  filter(n >= 4) %>% pull(primary)
mofaData <- mofaData[,useSamples]
```

```{r}
MOFAobject <- create_mofa_from_MultiAssayExperiment(mofaData)
MOFAobject
plot_data_overview(MOFAobject)
```

```{r, eval=FALSE}
MOFAobject_res <- running_mofa(obj=MOFAobject, var=0.01, mode="fast")
save(MOFAobject_res, file = "./output/MOFAres_contrast.RData")
```

Load result and show factors per view
```{r}
load("./output/MOFAres_contrast.RData")
plot_variance_explained(MOFAobject_res, x="view", y="factor")
```

```{r}
plot_variance_explained(MOFAobject_res, x="view", y="factor")
```

Factor correlation matrix
```{r, fig.width=5, fig.height=5}
plot_factor_cor(MOFAobject_res)
```

Total variance explained
```{r, fig.width=5, fig.height=4}
plot_variance_explained(MOFAobject_res, plot_total = T)[[2]]
```

## Factor heatmap

```{r, fig.height=6, fig.width=6}
library(pheatmap)
#gene annotation
facMat <- t(get_factors(MOFAobject_res)[[1]])[1:10,]

colAnno <- tibble(Name = colnames(facMat)) %>%
    mutate(IGHV = geneTab[Name, "IGHV.status"],
           trisomy12 = geneTab[Name, "trisomy12"],
           TP53 = geneTab[Name, "TP53"]) %>%
    data.frame() %>% column_to_rownames("Name")

pheatmap(facMat, clustering_method = "complete", annotation_col = colAnno,
         scale = "none")
```

Column medians vs factors of Ibrutinib
```{r}
fac <- facMat[1,]
# Vectors need to have the same length
facNut <- fac[names(fac) %in% colnames(exprObj$Ibrutinib)]
# Order Nutlin columns by factor names
medCol <- colMedians(exprObj$Ibrutinib[,names(facNut)])
plot(facNut, medCol)
```

Weight of genomic features on LF1
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 1,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

Weight of genomic features on LF7
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 7,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

Weight of genomic features on LF8
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 8,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

Weight of genomic features on LF6
```{r}
plot_top_weights(MOFAobject_res,
 view = "Gene",
 factor = 6,
 nfeatures = 10,     # Top number of features to highlight
 scale = T           # Scale weights from -1 to 1
)
```

# Session Info Details
```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```

