---
title: "Processing the untrimmed Salmon counted RNAseq of batches pilot and 1-4"
author: "Caroline Lohoff (based on code from Junyan Lu)"
date: "February 07 2022"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Description
The aim is to process RNA-Sequencing data from patients with chronic lymphocytic leukemia (CLL). The untrimmed RNA-Seq reads after drug perturbation were pre-processed with [Salmon](https://www.nature.com/articles/nmeth.4197). In comparison to a classical alignment approach such as [STAR](https://doi.org/10.1093/bioinformatics/bts635), Salmon combines a dual-phase parallel inference algorithm and feature-rich bias models with an ultra-fast read mapping procedure. This RMD file gathers the quantification files (quant.sf) created by Salmon, which are coming from the batches pilot, batch 1 combined, batch 2, batch 3, and batch 4. The read counts together with patient annotations are saved in a DESeq2 object. 

# Set options and load packages

Set global options
```{r setup, include=FALSE, cache = FALSE}
knitr::opts_knit$set(root.dir = "~/Documents/R/drugseq_test")
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

Load packages
```{r}
suppressPackageStartupMessages({
  library(DESeq2)
  library(tximport)
  library(EnsDb.Hsapiens.v86)
  library(tidyverse)
})
```

# Assemble Salmon results at gene level

Get transcript annotation using EnsDb
```{r}
geneAnno <- transcripts(EnsDb.Hsapiens.v86,
                        columns = c("tx_name","gene_id", "seq_name",
                                    "gene_name","gene_biotype"),
                        return.type = "DataFrame") %>% as_tibble()
geneAnno$tx_id <- NULL
```

Read in count tables of pilot batch, batch 1 combined, and batches 2-4
```{r}
rawFolder = "~/Documents/drugSeq/salmon/pilot_batch1comb_batch2_batch3_batch4/"
files <- list.files(rawFolder, recursive = TRUE, pattern = "quant.sf")
fileName <-  str_split(files, "[/]", simplify = TRUE)[,1]
names(files) = fileName
txi.salmon <- tximport(paste0(rawFolder,files), type = "salmon",
                       tx2gene = geneAnno, ignoreTxVersion = TRUE)
```

# Load patient annotations
```{r}
# Read in annotation file
patAnno <- readxl::read_xlsx("./data/RNAseq_samples_pilot_batch1-4_20220207.xlsx") %>%
  mutate(fileName = str_replace(fileName, ".txt.gz","")) %>%
  mutate(ID = paste0("smp",seq(nrow(.)))) %>%   # add column "ID" containing "smp1"-"smpn"
  mutate(Replicate = as.factor(Replicate)) %>%  # replicate numbers should be factors
  dplyr::select(-sequence) %>%                  # remove column "sequence"
  data.frame(stringsAsFactors = FALSE) %>%
  column_to_rownames("fileName")

order_vec <- patAnno$ID

# Create subset of data frame and add column to patAnno
#patAnno_small <- patAnno[c("fileName")]
#patAnno_small <- arrange(patAnno_small, fileName) %>%
#  mutate(IDplate = paste0("fn",seq(nrow(.))))

#patAnno <- patAnno %>%
#  left_join(patAnno_small, by = "fileName") %>%
#  column_to_rownames("fileName")

# Check all samples have annotations
all(fileName %in% rownames(patAnno))
all(rownames(patAnno) %in% fileName)

# Reorder by file names to match count tables
patAnno <- patAnno[fileName,]
```

# Assemble DEseq object
```{r}
ddsDrug <- DESeqDataSetFromTximport(txi.salmon,
                                    patAnno, design = ~1)
colnames(ddsDrug) <- ddsDrug$ID

# Annotate transcripts
rowAnno <- geneAnno %>%
  distinct(gene_id, seq_name, gene_name, gene_biotype) %>%
  dplyr::rename(chromosome=seq_name, symbol=gene_name, biotype=gene_biotype) %>%
  data.frame(stringsAsFactors = FALSE) %>%
  column_to_rownames("gene_id")
rowData(ddsDrug) <- rowAnno[rownames(ddsDrug),]

# Calculate size factors
ddsDrug <- estimateSizeFactors(ddsDrug)
```

Order rows and columns of ddsDrug object
```{r}
ddsDrug <- ddsDrug[,order_vec]
```

Save results
```{r}
save(ddsDrug, file = "./output/ddsDrug_untrimmed_batchPilot1c234_20220513.RData")
```

# Session Info Details
```{r, echo=FALSE, eval=TRUE}
sessionInfo()
```
