---
title: "Investigate the genomic regions the reads were aligned to by STAR "
author: "Junyan Lu"
date: "2021 Nov. 1"
output:
  BiocStyle::html_document:
    toc_float: true
---

```{r}
library(tidyverse)
```


# Plate annotations

```{r}
patAnno <- readxl::read_xlsx("../../data/drugSeq/RNAseq_samples_all.xlsx") %>%
  mutate(fileName = str_replace(fileName, ".txt.gz","")) %>%
  mutate(Replicate = as.factor(Replicate)) %>%
  select(-sequence) %>%
  mutate(cell = ifelse(patID == "Jurkat","Jurkat","Primary")) %>%
  select(fileName, patID, sampleID, batch, cell, treatment, FSC.SSC) %>%
  mutate_if(is.character, as.factor) %>%
  mutate(FSC.SSC = replace_na(FSC.SSC, 100))
```

# Filter bam files for the ones with baseline or DMSO

## Pilot
```{r}
allBam <- read_lines("../../scripts/bamList_pilot.txt")
allBamTab <- tibble(fullname = allBam) %>%
  mutate(filename = str_remove(fullname, "Aligned.sortedByCoord.out.bam")) %>%
  mutate(treatment = patAnno[match(filename, patAnno$fileName),]$treatment) %>%
  filter(treatment %in% c("Baseline","DMSO"))

write_lines(allBamTab$fullname, "../../scripts/bamList_pilot_DMSO.txt")
```


## Batch1
```{r}
allBam <- read_lines("../../scripts/bamList_batch1.txt")
allBamTab <- tibble(fullname = allBam) %>%
  mutate(filename = str_remove(fullname, "Aligned.sortedByCoord.out.bam")) %>%
  mutate(treatment = patAnno[match(filename, patAnno$fileName),]$treatment) %>%
  filter(treatment %in% c("Baseline","DMSO"))

write_lines(allBamTab$fullname, "../../scripts/bamList_batch1_DMSO.txt")
```


# Quality information from STAR alignment (genome mapping)

Process file list
```{r, warning=FALSE, error=FALSE, eval=FALSE}
allFiles <- list.files("../../data/drugSeq/calcCov/",recursive = TRUE, full.names = TRUE, pattern = ".out")

fileTab <- tibble(fullName = allFiles) %>%
  mutate(baseName = basename(fullName),
         filename = str_remove(baseName, "Aligned_(exon|intron|intergenetic).out")) %>%
  mutate(region = case_when(str_detect(baseName,"exon") ~ "exon",
                              str_detect(baseName,"intron") ~ "intron",
                              TRUE ~ "intergenetic"))

fileTab <- left_join(fileTab, patAnno, by = c(filename = "fileName"))

#only keep baseline or DMSO samples
fileTab <- filter(fileTab, treatment %in% c("DMSO","Baseline"))

#Remove incomplete record
uniqueTab <- distinct(fileTab, filename, region) %>%
  group_by(filename) %>% summarise(n=length(region)) %>%
  filter(n==3)
fileTab <- filter(fileTab, filename %in% uniqueTab$filename)
```

Read record
```{r, eval=FALSE}
allRecord <- lapply(seq(nrow(fileTab)), function(i){
  rec <- fileTab[i,]
  read_tsv(rec$fullName, 
           col_names = c("chr","start","end","reads","covered","length","coverage"), 
           col_types = "ciiiiid") %>%
    select(chr, start, end, reads, coverage) %>%
    mutate(filename = rec$filename, region = rec$region) %>%
    filter(chr %in% c(as.character(seq(1,22)),"X","Y"))
    
}) %>% bind_rows() %>% 
  mutate(chr = factor(chr), region = factor(region), filename = factor(filename))
```

Post-processing
```{r, eval=FALSE}
allRecord <- left_join(allRecord, patAnno, by = c(filename = "fileName"))

allRecord <- group_by(allRecord,filename) %>%
  mutate(nTotal = sum(reads)) %>%
  ungroup() %>%
  mutate(normReads = reads/nTotal*mean(nTotal)) %>%
  select(-nTotal,-treatment)

save(allRecord, file = "allRecord.RData")
```

```{r}
load("allRecord.RData")
```

# Average reads
```{r}
totalRead <- group_by(allRecord,filename) %>%
  summarise(nTotal = sum(reads))

sumRead <- group_by(allRecord, filename, patID, batch, cell, region) %>%
  summarise(nRegion = sum(reads), FSC.SSC = mean(FSC.SSC,na.rm=TRUE)) %>%
  left_join(totalRead, by = "filename") %>%
  mutate(fraction = nRegion/nTotal)
```


```{r,fig.height=5, fig.width=10}
ggplot(sumRead, aes(x=batch, y=fraction)) +
  #geom_boxplot(outlier.shape = NA) + 
  ggbeeswarm::geom_quasirandom(aes(col = cell, alpha = FSC.SSC),size=4) +
  facet_wrap(~region) +
  ylim(0,1) +
  theme_bw()
```

# Average coverage
```{r}
sumCov <- group_by(allRecord, filename, patID, batch, cell, region) %>%
  summarise(avgCov = mean(coverage),FSC.SSC = mean(FSC.SSC,na.rm=TRUE))
```


```{r,fig.height=5, fig.width=10}
ggplot(sumCov, aes(x=batch, y=avgCov)) +
  geom_boxplot(outlier.shape = NA) + 
  ggbeeswarm::geom_quasirandom(aes(col = cell, alpha=FSC.SSC), size=3) +
  facet_wrap(~region) +
  #ylim(0,1) +
  theme_bw()
```


# Per-chromosome plot

## Normalized reads numbers
```{r}
allPlots <- lapply(sort(unique(allRecord$chr)), function(iChr) {
  
  plotTab <- filter(allRecord, chr == iChr, reads > 0) %>%
    arrange(start) %>% select(start, normReads, region, batch, cell, filename) %>%
    mutate(type = paste0(batch,"_",cell))
   
  #subsampleing
  allStart <- unique(plotTab$start)
  subStart <- sample(allStart, as.integer(length(allStart)*0.05))
  plotTab <- filter(plotTab, start %in% subStart)

  g <- ggplot(plotTab, aes(x=start, y = log10(normReads))) +
    geom_point(aes(col = region, shape = type)) +
    #xlim(0, max(plotTab$start)) +
    scale_x_continuous(expand = c(0,0)) +
    scale_color_manual(values = c(exon="salmon",intron = "lightblue",intergenetic = "grey50")) +
    scale_shape_manual(values = c(pilot_Primary = 16, batch1_Primary = 1, batch1_Jurkat = 17)) +
    theme_bw() + ggtitle(paste0("Chromosome: ", iChr)) +
    theme(plot.title = element_text(size=10, face = "bold"))
  
  g

})


jyluMisc::makepdf(allPlots, name = "reads_per_chom.pdf", ncol=1, nrow=1 ,figNum = 1, height = 8, width = 80)
```
[reads_per_chom.pdf](./reads_per_chom.pdf)
**Note that the reads are sub-sampled at 5% to avoid overcrowding**

## Coverage
```{r}
allPlots <- lapply(sort(unique(allRecord$chr)), function(iChr) {
  
  plotTab <- filter(allRecord, chr == iChr, coverage > 0) %>%
    arrange(start) %>% select(start, coverage, region, batch, cell, filename) %>%
    mutate(type = paste0(batch,"_",cell))
   
  #subsampleing
  allStart <- unique(plotTab$start)
  subStart <- sample(allStart, as.integer(length(allStart)*0.05))
  plotTab <- filter(plotTab, start %in% subStart)

  g <- ggplot(plotTab, aes(x=start, y = coverage)) +
    geom_point(aes(col = region, shape = type)) +
    #xlim(0, max(plotTab$start)) +
    scale_x_continuous(expand = c(0,0)) +
    scale_color_manual(values = c(exon="salmon",intron = "lightblue",intergenetic = "grey50")) +
    scale_shape_manual(values = c(pilot_Primary = 16, batch1_Primary = 1, batch1_Jurkat = 17)) +
    theme_bw() + ggtitle(paste0("Chromosome: ", iChr)) +
    theme(plot.title = element_text(size=10, face = "bold"))
  
  g

})


jyluMisc::makepdf(allPlots, name = "coverage_per_chom.pdf", ncol=1, nrow=1 ,figNum = 1, height = 8, width = 80)
```
[coverage_per_chom.pdf](coverage_per_chom.pdf)  
**Note that the reads are sub-sampled at 5% to avoid overcrowding**

# Per-chromosome plot (intergenetic only)

## Normalized read numbers
```{r}
allPlots <- lapply(sort(unique(allRecord$chr)), function(iChr) {
  
  plotTab <- filter(allRecord, chr == iChr, reads > 0, region == "intergenetic") %>%
    arrange(start) %>% select(start, normReads, region, batch, cell, filename) %>%
    mutate(type = paste0(batch,"_",cell))
   
  #subsampleing
  allStart <- unique(plotTab$start)
  subStart <- sample(allStart, as.integer(length(allStart)*1))
  plotTab <- filter(plotTab, start %in% subStart)

  g <- ggplot(plotTab, aes(x=start, y = log10(normReads))) +
    geom_point(aes(col = type)) +
    #xlim(0, max(plotTab$start)) +
    scale_x_continuous(expand = c(0,0)) +
    #scale_color_manual(values = c(exon="salmon",intron = "lightblue",intergenetic = "grey50")) +
    scale_color_manual(values = c(pilot_Primary = "red", batch1_Primary = "grey50", batch1_Jurkat = "Orange")) +
    theme_bw() + ggtitle(paste0("Chromosome: ", iChr)) +
    theme(plot.title = element_text(size=10, face = "bold"))
  
  g

})


jyluMisc::makepdf(allPlots, name = "intergenetic_reads_per_chom.pdf", ncol=1, nrow=1 ,figNum = 1, height = 8, width = 80)
```
[intergenetic_reads_per_chom.pdf](./intergenetic_reads_per_chom.pdf)   
**All reads are considered**

## Coverage
```{r}
allPlots <- lapply(sort(unique(allRecord$chr)), function(iChr) {
  
  plotTab <- filter(allRecord, chr == iChr, coverage > 0, region == "intergenetic") %>%
    arrange(start) %>% select(start, coverage, region, batch, cell, filename) %>%
    mutate(type = paste0(batch,"_",cell))
   
  #subsampleing
  allStart <- unique(plotTab$start)
  subStart <- sample(allStart, as.integer(length(allStart)*1))
  plotTab <- filter(plotTab, start %in% subStart)

  g <- ggplot(plotTab, aes(x=start, y = coverage)) +
    geom_point(aes(col = type)) +
    #xlim(0, max(plotTab$start)) +
    scale_x_continuous(expand = c(0,0)) +
    scale_color_manual(values = c(pilot_Primary = "red", batch1_Primary = "grey50", batch1_Jurkat = "Orange")) +
    #scale_color_manual(values = c(exon="salmon",intron = "lightblue",intergenetic = "grey50")) +
    #scale_shape_manual(values = c(pilot_Primary = 16, batch1_Primary = 1, batch1_Jurkat = 17)) +
    theme_bw() + ggtitle(paste0("Chromosome: ", iChr)) +
    theme(plot.title = element_text(size=10, face = "bold"))
  
  g

})


jyluMisc::makepdf(allPlots, name = "intergenetic_coverage_per_chom.pdf", ncol=1, nrow=1 ,figNum = 1, height = 8, width = 80)
```
[intergenetic_coverage_per_chom.pdf](./intergenetic_coverage_per_chom.pdf)   
**All reads are considered**
