---
title: "Processing the Salmon counted RNAseq"
author: "Junyan Lu"
date: "2/25/2019"
output:
  BiocStyle::html_document:
    toc_float: true
---

# Load libraries

Libraries
```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(tximport)
library(EnsDb.Hsapiens.v86)
library(RColorBrewer)
library(pheatmap)
library(geneplotter)
library(tidyverse)
```

Global variables
```{r}
#set the global ggplot theme
theme_set(theme_bw() + theme(axis.text = element_text(size=12), 
                             axis.title = element_text(size=14),
                             plot.title = element_text(size = 15, hjust =0.5, face="bold")))
```

# Assemble Salmon results at gene level

Get transcript annotation using EnsDb
```{r}
geneAnno <- transcripts(EnsDb.Hsapiens.v86, columns = c("tx_name","gene_id", "seq_name",
                                                        "gene_name","gene_biotype"), return.type = "DataFrame") %>% as_tibble()
geneAnno$tx_id <- NULL
```

Read in count tables (aligned by Salmon)
```{r}
rawFolder = "/g/huber/projects/nct/cll/ProcessedData/RNAseq/drugSeq/salmon_trimmed/"
files <- list.files(rawFolder, recursive = TRUE, pattern = "quant.sf")
fileName <-  str_split(files, "[/]", simplify = TRUE)[,2]
names(files) = fileName
txi.salmon <- tximport(paste0(rawFolder,files), type = "salmon", tx2gene = geneAnno, ignoreTxVersion = TRUE)
```


Annotate patients
```{r}
patAnno <- readxl::read_xlsx("../../data/drugSeq/RNAseq_samples_all.xlsx") %>%
  mutate(fileName = str_replace(fileName, ".txt.gz","")) %>%
  mutate(ID = paste0("smp",seq(nrow(.)))) %>%
  mutate(Replicate = as.factor(Replicate)) %>%
  select(-sequence) %>%
  data.frame(stringsAsFactors = FALSE) %>%
  column_to_rownames("fileName") 

#check all samples have annotations
all(fileName %in% rownames(patAnno))
all(rownames(patAnno) %in% fileName)

#reorder 
patAnno <- patAnno[fileName,]
```


Assemble DEseq object
```{r}
ddsDrug <- DESeqDataSetFromTximport(txi.salmon, patAnno, design = ~1)
colnames(ddsDrug) <- ddsDrug$ID

#annotate transcripts
rowAnno <- geneAnno %>% distinct(gene_id, seq_name, gene_name, gene_biotype) %>%
  dplyr::rename(chromosome = seq_name, symbol = gene_name, biotype = gene_biotype) %>%
  data.frame(stringsAsFactors = FALSE) %>% column_to_rownames("gene_id")
rowData(ddsDrug) <- rowAnno[rownames(ddsDrug),]

ddsDrug <- estimateSizeFactors(ddsDrug)
```


# Save results
```{r}
save(ddsDrug,  file = "../../var/ddsDrug_trimmed_20211011.RData")
```

