### Scripts for preprocessing data:

__process.Rmd__  script for processing the data generated by Salmmon

__process_htseq.Rmd__ script for processing the data generated by the STAR+HTseq workflow

### Scripts for quality control

__bamQC*.Rmd__ files contain the codes I used to process and visualise the quality reports from STAR, HTseq and Salmon, to investigate the difference of QC metrics for different batches.   

__calcCoverage.Rmd__ was used to estimate the distributions of the genomic regions where the RNAseq reads were mapped to, in order to see whether the poor mapping rate of some samples was becasue the reads were mapped to intergenetic regions.   

__qualityCheck*.Rmd__ files were used to check the quality of RNAseq count data, processed with different workflows (Salmon, Salmon + reads trimming, STAR+htseq)


### Scripts for differential expression

__differentialExpression*.Rmd__ scripts for differential analysis based on the output from Salmon or STAR+HTseq.




