# Generate script for submitting STAR alignment job to EMLB cluster

# Both fastq files of batch 1 and batch 1 new will be combined

library(tidyverse)

inputList = "~/Documents/R/drugseq_test/data/fastqFiles_batch1.txt"

rawFolder1 = "/g/huber/projects/nct/cll/RawData/RNASeq/drugSeq/batch1"
rawFolder2 = "/g/huber/projects/nct/cll/RawData/RNASeq/drugSeq/batch1_new"

refGenome = "/g/huber/projects/nct/cll/RawData/RNASeq/ReferenceGenome/GRCh38_Ensembl104/star2.7.9a/"

runTime = "01:30:00" #two hours
nThread = 4  #4 cores per node
nJob = 10 #number of jobs in each script

### Cluster Usage Report ###
#Nodes: 1
#Cores per node: 4
#CPU Utilized: 01:30:02
#CPU Efficiency: 39.09% of 03:50:20 core-walltime
#Job Wall-clock time: 00:57:35
#Memory Utilized: 28.28 GB
#Memory Efficiency: 88.37% of 32.00 GB


#function to change file name between the two runs
fixName <- function(x) {
  x_new <- str_replace(x, "HMCKNBBXY", "HMC5GBBXY") %>%
    str_replace("-1-1_Lu", "-1-2_Lu") %>%
    str_replace("lane812dot2P", "lane8122P")
}

#read in file list
fastq <- readLines(inputList)

#function to generate the header for script
genHeader <- function(i, email = FALSE) {
  header <- c("#!/bin/bash",
            sprintf("#SBATCH -J align_%s",i),
            sprintf("#SBATCH -t %s",runTime),
            sprintf("#SBATCH -n %s",nThread),
            "#SBATCH --mem=32GB",
            sprintf("#SBATCH -o slurm_%s.out",i),
            sprintf("#SBATCH -e slurm_%s.err",i))
  if (email == TRUE) {
    header <- c(header,
                "#SBATCH --mail-type=END,FAIL",
                "#SBATCH --mail-user=caroline.lohoff@embl.de")
  }

  header = c(header,"","module load STAR/2.7.9a-GCC-10.3.0") #load module
  return(header)
}


#split jobs
jobList <- split(fastq, ceiling(seq_along(fastq)/nJob))

#generate script for each run
for (i in seq_along(jobList)) {
  fileConn<-file(sprintf("sub%s.sh",i))
  patID <- sapply(jobList[[i]], function(x) str_split(x,"[.]")[[1]][1])
  starCommand <- sapply(seq_along(jobList[[i]]), function(j) {
    f1 <- jobList[[i]][[j]]
    f2 <- fixName(jobList[[i]][[j]])
    c("",
      sprintf("cp %s/%s ${TMPDIR}", rawFolder1, f1),
      sprintf("cp %s/%s ${TMPDIR}", rawFolder2, f2),
      "cd $TMPDIR",
      sprintf("STAR --runThreadN %s --outBAMsortingThreadN %s --genomeDir %s --readFilesIn %s,%s --outFileNamePrefix %s --outSAMtype BAM SortedByCoordinate --readFilesCommand zcat",
      nThread, nThread, refGenome, f1, f2, patID[j]),
      sprintf("mv *.bam ${SLURM_SUBMIT_DIR}"),
      sprintf("mv *Log.final.out ${SLURM_SUBMIT_DIR}/log/"),
      sprintf("rm %s", jobList[[i]][j]),
      sprintf("rm %s", fixName(jobList[[i]][[j]])),
      "")
  })
  writeLines(c(genHeader(i),starCommand), con = fileConn)
  close(fileConn)
}

#generate script for submitting all jobs
fileConn <- file("subAll.sh")
allSub <- sapply(seq_along(jobList), function(i) {
  sprintf("sbatch sub%s.sh",i)
})
writeLines(allSub, fileConn)
close(fileConn)
